---
title: "b.subtilis_etc"
author: "Mikhail Orlov"
date: "16.10.16"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Libraries}

rm(list = ls())
library(seqinr)
library(fastcluster)
library(reldna)
library(dendextend)
library(ggplot2)
library(ape)
library(RColorBrewer)
```

```{r Reading in fastas from Umarov&Solovyev, 2017}

e.coli_proms <- read.fasta('https://raw.githubusercontent.com/solovictor/CNNPromoterData/master/Ecoli_prom.fa')

e.coli_non_proms <- read.fasta('https://raw.githubusercontent.com/solovictor/CNNPromoterData/master/Ecoli_non_prom.fa')

e.coli_proms_non_proms <- c(e.coli_proms, e.coli_non_proms)

b.subtilis_proms <- read.fasta('https://raw.githubusercontent.com/solovictor/CNNPromoterData/master/Bacillus_prom.fa')

b.subtilis_non_proms <- read.fasta('https://raw.githubusercontent.com/solovictor/CNNPromoterData/master/Bacillus_non_prom.fa')

b.subtilis_proms_non_proms <- c(b.subtilis_proms, b.subtilis_non_proms)


seqs <- grep('*_proms', ls(), value = T)

sapply(seqs, function(x) {
  sapply(x, FUN = function(x) {length(get(x))})
})

```

```{r Choosing colors for 4 type of sequences}
library(RColorBrewer)
cols_brew <- RColorBrewer::brewer.pal(4, 'Dark2')
plot(seq_along(cols_brew), col = cols_brew, lwd =25)
e.coli_proms_col <- cols_brew[1]
e.coli_non_proms_col <- cols_brew[2]
b.subtilis_proms_col <- cols_brew[3]
b.subtilis_non_proms_col <- cols_brew[4]

```



```{r sequences physics, function for open states dynamics calculation [Grinevich et al., 2015]}

dynchars<-function(seq, interval_size) {
  if (missing(seq))
    stop("Need to specify sequence (as a vector of chars)")
  
  if (missing(interval_size))
    stop("Need to specify interval size")
  
  if(!is.character(seq))
    stop("Sequence must be a character vector containing A, C, G, T letters only")
  
seq<-toupper(seq)
seq<-c(seq, seq[2:(interval_size)])
  
a<-3.4*10^(-10)
I<-c(7.6, 4.8, 8.2, 4.1)*10^(-44)
K<-c(227, 155, 220, 149)*10^(-20)
V<-c(2.09, 1.43, 3.12, 2.12)*10^(-20)
tau<-c(127, 99, 140, 84)
  
csA<-cumsum(seq=='A') 
csT<-cumsum(seq=='T')
csG<-cumsum(seq=='G')
csC<-cumsum(seq=='C')

countA = csA[interval_size:length(csA)]-c(0, csA[1:(length(csA)-interval_size)])
countT = csT[interval_size:length(csT)]-c(0, csT[1:(length(csT)-interval_size)])
countG = csG[interval_size:length(csG)]-c(0, csG[1:(length(csG)-interval_size)])
countC = csC[interval_size:length(csC)]-c(0, csC[1:(length(csC)-interval_size)])
  
M<-cbind(countA, countT, countG, countC)/interval_size
M_comp<-cbind(countT, countA, countC, countG)/interval_size
M_comp<-apply(t(M_comp),1,rev) 
Is<-as.numeric(M%*%I)#! numeric conversion
Ks<-as.numeric(M%*%K)
Vs<-as.numeric(M%*%V)
  
E01<-(8*(Ks*Vs)^0.5)* 6E23 / 4184
d1<-((Ks*a^2)/Vs)^(0.5)/a;
c1<-(Ks*a^2/Is)^0.5
m1<-E01/c1/6.011E-26
taus1<-as.numeric(M%*%tau) #!as.numeric conversion
gc1 = M[,3] + M[,4]
  
Is<-as.numeric(M%*%I)#! numeric conversion
Ks<-as.numeric(M%*%K)
Vs<-as.numeric(M%*%V)

  
E02<- 8*(Ks*Vs)^0.5  * 6E23 / 4184;
d2<-((Ks*a^2)/Vs)^(0.5)/a;
c2<-(Ks*a^2/Is)^0.5;
m2<-E02/c2/6.011E-26;
taus2<-as.numeric(M_comp%*%tau)
gc2 = M_comp[,3] + M_comp[,4]

dynchars_return<-list(E01=E01, d1=d1, c1=c1, m1=m1, taus1=taus1, gc1=gc1, E02=E02, d2=d2, c2=c2, m2=m2, taus2=taus2, gc2=gc2)
  
return(dynchars_return)
  
}

```

```{r}
e.coli_proms_dynchars <- lapply(e.coli_proms, function(x) {dynchars(x, interval_size = 25)})
e.coli_non_proms_dynchars <- lapply(e.coli_non_proms, function(x) {dynchars(x, interval_size = 25)})

b.subtilis_proms_dynchars <- lapply(b.subtilis_proms, function(x) {dynchars(x, interval_size = 25)})
b.subtilis_non_proms_dynchars <- lapply(b.subtilis_non_proms, function(x) {dynchars(x, interval_size = 25)})

e.coli_proms_E01 <- sapply(e.coli_proms_dynchars, function(x) {x$E01})
e.coli_non_proms_E01 <- sapply(e.coli_non_proms_dynchars, function(x) {x$E01})

b.subtilis_proms_E01 <- sapply(b.subtilis_proms_dynchars, function(x) {x$E01})
b.subtilis_non_proms_E01 <- sapply(b.subtilis_non_proms_dynchars, function(x) {x$E01})
#some statistics
par(mfrow = c (2,2))
ylim = c(170, 290)
boxplot(as.numeric(e.coli_proms_E01), ylim = ylim, main = 'E0, E. coli promoters')
boxplot(as.numeric(e.coli_non_proms_E01), ylim = ylim, main = 'E0, E. coli non_promoters')

boxplot(as.numeric(b.subtilis_proms_E01), ylim = ylim, main = 'E0, B. subtilis promoters')
boxplot(as.numeric(b.subtilis_non_proms_E01), ylim = ylim, main = 'E0, B. subtilis non-promoters')
```

```{r E01}
#par(mfrow = c(2,2),
 #   mar = rep(0.7,4))
matplot(sapply(e.coli_proms_dynchars, function(x) {x$E01}), type = 'l', col = 1, lwd = .1, ylim = ylim, main = 'E0, E. coli promoters')

matplot(sapply(e.coli_non_proms_dynchars, function(x) {x$E01}), type = 'l', col = 1, lwd = .1, ylim = ylim, main = 'E0, E. coli non_promoters')

matplot(sapply(b.subtilis_proms_dynchars, function(x) {x$E01}), type = 'l', col = 1, lwd = .1, ylim = ylim, main = 'E0, B. subtilis promoters')

matplot(sapply(b.subtilis_non_proms_dynchars, function(x) {x$E01}), type = 'l', col = 1, lwd = .1, ylim = ylim, main = 'E0, B. subtilis non-promoters')
```


```{r gc}

#par(mfrow = c(2,2),
  #  mar = rep(0.7,4))
ylim = c(0,1)
matplot(sapply(e.coli_proms_dynchars, function(x) {x$gc1}), type = 'l', col = 1, lwd = .3, ylim = ylim, main = 'GC, E. coli promoters')
matplot(sapply(e.coli_non_proms_dynchars, function(x) {x$gc1}), type = 'l', col = 1, lwd = .3, ylim = ylim, main = 'GC, E. coli non-promoters')

matplot(sapply(b.subtilis_proms_dynchars, function(x) {x$gc1}), type = 'l', col = 1, lwd = .3, ylim = ylim, main = 'GC, B subtilis promoters')
matplot(sapply(b.subtilis_non_proms_dynchars, function(x) {x$gc1}), type = 'l', col = 1, lwd = .3, ylim = ylim, main = 'GC, B subtilis non-promoters')
```

```{r EP}

X <- -130:30
e.coli_proms_mpots <- sapply(e.coli_proms, FUN = function(x) {
  tmp <- lseqspline1D(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1, ref = 41)
  (tmp$mpot[which(tmp$x %in% X)])
})
e.coli_non_proms_mpots <- sapply(e.coli_non_proms, FUN = function(x) {
  tmp <- lseqspline1D(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1, ref = 41)
  (tmp$mpot[which(tmp$x %in% X)])
})

b.subtilis_proms_mpots <- sapply(b.subtilis_proms, FUN = function(x) {
  tmp <- lseqspline1D(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1, ref = 41)
  (tmp$mpot[which(tmp$x %in% X)])
})
b.subtilis_non_proms_mpots <- sapply(b.subtilis_non_proms, FUN = function(x) {
  tmp <- lseqspline1D(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1, ref = 41)
  (tmp$mpot[which(tmp$x %in% X)])
})

##par(mfrow = c(2,2))

matplot(e.coli_proms_mpots[-c(0:60),], type = 'l', col = 1, lwd = .4, main = 'EP, E. coli promoters')
matplot(e.coli_non_proms_mpots[-c(0:60),], type = 'l', col = 1, lwd = .4, main = 'EP, E. coli non-promoters')
matplot(b.subtilis_proms_mpots[-c(0:60),], type = 'l', col = 1, lwd = .4, main = 'EP, B. subtilis promoters')
matplot(b.subtilis_non_proms_mpots[-c(0:60),], type = 'l', col = 1, lwd = .4, main = 'EP, B. subtilis non-promoters')
```

```{r bendability}

X <- -130:30
e.coli_proms_bends <- sapply(e.coli_proms, FUN = function(x) {
  tmp <- bendability(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1)
})

e.coli_non_proms_bends <- sapply(e.coli_non_proms, FUN = function(x) {
  tmp <- bendability(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1)
})

b.subtilis_proms_bends <- sapply(b.subtilis_proms, FUN = function(x) {
    tmp <- bendability(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1)
})

b.subtilis_non_proms_bends <- sapply(b.subtilis_non_proms, FUN = function(x) {
     tmp <- bendability(s = paste0(x, collapse = ''), bound = c(1, length(x)), width = 1)
})

##par(mfrow = c(2,2))
lwd = 0.2
matplot(e.coli_proms_bends, type = 'l', col = 1, lwd = lwd, main = 'Bendability, E. coli promoters')
matplot(e.coli_non_proms_bends, type = 'l', col = 1, lwd = lwd, main = 'Bendability, E. coli non-promoters')

matplot(b.subtilis_proms_bends, type = 'l', col = 1, lwd = lwd, main = 'Bendability, B. subtilis promoters')
matplot(b.subtilis_non_proms_bends, type = 'l', col = 1, lwd = lwd, main = 'Bendability, B. subtilis non-promoters')

ward_e.coli_proms_bends <- hclust.vector(t(e.coli_proms_bends), method = 'ward')
plot(ward_e.coli_proms_bends, labels = F, main = 'Ward clusterization, bendability for E. coli promoters')

k=3
cut_e.coli_proms_bends <- cutree(ward_e.coli_proms_bends, k = k)

par(mfrow = c (1,3))
for (i in unique(cut_e.coli_proms_bends)) {
  no <- which(cut_e.coli_proms_bends == i)
  matplot(e.coli_proms_bends[,no], type = 'l', col = 1, lwd = 0.1, ylim = c(-1.5, 0.75), main = paste0('Bendability, cluster ', i))
}

```

```{r promoters vs non-promoters for each bacteria2}
e.coli_proms_non_proms_E01 <- t(cbind(sapply(e.coli_proms_dynchars, function(x) {x$E01}),
      sapply(e.coli_non_proms_dynchars, function(x) {x$E01})))

ward_e.coli_proms_non_proms_E01 <- hclust.vector(e.coli_proms_non_proms_E01, method = 'ward')
plot(ward_e.coli_proms_non_proms_E01, labels = F, main = 'E.coli promoters and non-promoters (E0)')



colvec_E01 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(e.coli_non_proms_col, length(e.coli_non_proms_dynchars)))[order.dendrogram(ward_e.coli_proms_non_proms_E01 %>% as.dendrogram )]
ward_e.coli_proms_non_proms_E01 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_E01$labels), col = colvec_E01) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_E01$height)*.2) -> to_plot_e.coli_proms_non_proms_E01


e.coli_proms_non_proms_gc1 <- t(cbind(sapply(e.coli_proms_dynchars, function(x) {x$gc1}),
      sapply(e.coli_non_proms_dynchars, function(x) {x$gc1})))

ward_e.coli_proms_non_proms_gc1 <- hclust.vector(e.coli_proms_non_proms_gc1, method = 'ward')
plot(ward_e.coli_proms_non_proms_gc1, labels = F, main = 'E.coli promoters and non-promoters (GC)')



colvec_gc1 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(e.coli_non_proms_col, length(e.coli_non_proms_dynchars)))[order.dendrogram(ward_e.coli_proms_non_proms_gc1 %>% as.dendrogram )]
ward_e.coli_proms_non_proms_gc1 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_gc1$labels), col = colvec_gc1) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_gc1$height)*.2) -> to_plot_e.coli_proms_non_proms_gc1


#par(mfrow = c(2,1))
#plot(to_plot_e.coli_proms_non_proms_E01, main ='E. coli promoter and non-promoter mixture (E01)')
#plot(to_plot_e.coli_proms_non_proms_gc1, main ='E. coli promoter and non-promoter mixture (GC)')
#legend('top', lty = 1, col = c(e.coli_proms_col, e.coli_non_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')
```

```{r promoters vs non-promoters for each bacteria - B.subtilis 2}
b.subtilis_proms_non_proms_E01 <- t(cbind(sapply(b.subtilis_proms_dynchars, function(x) {x$E01}),
      sapply(b.subtilis_non_proms_dynchars, function(x) {x$E01})))

ward_b.subtilis_proms_non_proms_E01 <- hclust.vector(b.subtilis_proms_non_proms_E01, method = 'ward')
plot(ward_b.subtilis_proms_non_proms_E01, labels = F, main = 'B. subtilis promoters and non-promoters (E0)')



colvec_E01 <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms_dynchars)))[order.dendrogram(ward_b.subtilis_proms_non_proms_E01 %>% as.dendrogram )]
ward_b.subtilis_proms_non_proms_E01 %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_E01$labels), col = colvec_E01) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_E01$height)*.2) -> to_plot_b.subtilis_proms_non_proms_E01


b.subtilis_proms_non_proms_gc1 <- t(cbind(sapply(b.subtilis_proms_dynchars, function(x) {x$gc1}),
      sapply(b.subtilis_non_proms_dynchars, function(x) {x$gc1})))

ward_b.subtilis_proms_non_proms_gc1 <- hclust.vector(b.subtilis_proms_non_proms_gc1, method = 'ward')
#plot(to_plot_b.subtilis_proms_non_proms_E01, labels = F, main = 'B. subtilis promoters and non-promoters (GC)')



colvec_gc1 <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms_dynchars)))[order.dendrogram(ward_b.subtilis_proms_non_proms_gc1 %>% as.dendrogram )]
ward_b.subtilis_proms_non_proms_gc1 %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_gc1$labels), col = colvec_gc1) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_gc1$height)*.2) -> to_plot_b.subtilis_proms_non_proms_gc1

#par(mfrow = c(2,1))
#plot(to_plot_b.subtilis_proms_non_proms_E01, main ='B. subtilis promoter and non-promoter mixture (E01)')
#plot(to_plot_b.subtilis_proms_non_proms_gc1, main ='B. subtilis promoter and non-promoter mixture (GC)')
#legend('top', lty = 1, col = c(b.subtilis_proms_col, b.subtilis_non_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')
```

```{r promoters vs non-promoters for each bacteria1}

library(ape)
e.coli_proms_non_proms_seqs <- t(cbind(sapply(e.coli_proms, function(x) {unlist(x)}),
      sapply(e.coli_non_proms, function(x) {unlist(x)})))


e.coli_proms_non_proms_dnadist <- ape::dist.dna(e.coli_proms_non_proms_seqs, model = 'raw', pairwise.deletion = T)

#njs <- (njs(e.coli_proms_non_proms_dnadist ) )

#plot(njs)
ward_e.coli_proms_non_proms_E01 <- hclust.vector(e.coli_proms_non_proms_E01, method = 'ward')




colvec_E01 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(e.coli_non_proms_col, length(e.coli_non_proms_dynchars)))[order.dendrogram(ward_e.coli_proms_non_proms_E01 %>% as.dendrogram )]
ward_e.coli_proms_non_proms_E01 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_E01$labels), col = colvec_E01) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_E01$height)*.2) -> to_plot_E01


e.coli_proms_non_proms_gc1 <- t(cbind(sapply(e.coli_proms_dynchars, function(x) {x$gc1}),
      sapply(e.coli_non_proms_dynchars, function(x) {x$gc1})))

ward_e.coli_proms_non_proms_gc1 <- hclust.vector(e.coli_proms_non_proms_gc1, method = 'ward')

colvec_gc1 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(e.coli_non_proms_col, length(e.coli_non_proms_dynchars)))[order.dendrogram(ward_e.coli_proms_non_proms_gc1 %>% as.dendrogram )]

#
ward_e.coli_proms_non_proms_gc1 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_gc1$labels), col = colvec_gc1) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_gc1$height)*.2) -> to_plot_gc1

#par(mfrow = c(2,1))
#plot(to_plot_E01, main ='E. coli promoter and non-promoter mixture (E01)')
#plot(to_plot_gc1, main ='E. coli promoter and non-promoter mixture (GC)')
#legend('top', lty = 1, col = c(e.coli_proms_col, e.coli_non_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')
```

```{r promoters vs non-promoters for each bacteria - B.subtilis 1}
b.subtilis_proms_non_proms_E01 <- t(cbind(sapply(b.subtilis_proms_dynchars, function(x) {x$E01}),
      sapply(b.subtilis_non_proms_dynchars, function(x) {x$E01})))

ward_b.subtilis_proms_non_proms_E01 <- hclust.vector(b.subtilis_proms_non_proms_E01, method = 'ward')



colvec_E01 <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms_dynchars)))[order.dendrogram(ward_b.subtilis_proms_non_proms_E01 %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_b.subtilis_proms_non_proms_E01 %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_E01$labels), col = colvec_E01) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_E01$height)*.2) -> to_plot_E01


b.subtilis_proms_non_proms_gc1 <- t(cbind(sapply(b.subtilis_proms_dynchars, function(x) {x$gc1}),
      sapply(b.subtilis_non_proms_dynchars, function(x) {x$gc1})))

ward_b.subtilis_proms_non_proms_gc1 <- hclust.vector(b.subtilis_proms_non_proms_gc1, method = 'ward')

colvec_gc1 <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms_dynchars)))[order.dendrogram(ward_b.subtilis_proms_non_proms_gc1 %>% as.dendrogram )]

#
ward_b.subtilis_proms_non_proms_gc1 %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_gc1$labels), col = colvec_gc1) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_gc1$height)*.2) -> to_plot_gc1

#par(mfrow = c(2,1))
#plot(to_plot_E01, main ='B. subtilis promoter and non-promoter mixture (E01)')
#plot(to_plot_gc1, main ='B. subtilis promoter and non-promoter mixture (GC)')
#legend('top', lty = 1, col = c(b.subtilis_proms_col, b.subtilis_non_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')
```

```{r promoters vs non-promoters for both bacteria }
#e01
e.coli_b.subtilis_proms_E01 <- t(cbind(sapply(e.coli_proms_dynchars, function(x) {x$E01}),
      sapply(b.subtilis_proms_dynchars, function(x) {x$E01})))

ward_e.coli_b.subtilis_proms_E01<- hclust.vector(e.coli_b.subtilis_proms_E01, method = 'ward')

colvec_E01 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)))[order.dendrogram(ward_e.coli_b.subtilis_proms_E01 %>% as.dendrogram )]
ward_e.coli_b.subtilis_proms_E01 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_proms_E01$labels), col = colvec_E01) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_proms_E01$height)*.2) -> to_plot_e.coli_b.subtilis_proms_E01


#plot(to_plot_e.coli_b.subtilis_proms_E01, main ='E. coli and B. subtilis promoters: E01')
#legend('top', lty = 1, col = c(e.coli_proms_col, b.subtilis_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')



#gc1

e.coli_b.subtilis_proms_gc1 <- t(cbind(sapply(e.coli_proms_dynchars, function(x) {x$gc1}),
      sapply(b.subtilis_proms_dynchars, function(x) {x$gc1})))

ward_e.coli_b.subtilis_proms_gc1<- hclust.vector(e.coli_b.subtilis_proms_gc1, method = 'ward')

colvec_gc1 <- c(rep(e.coli_proms_col, length(e.coli_proms_dynchars)),rep(b.subtilis_proms_col, length(b.subtilis_proms_dynchars)))[order.dendrogram(ward_e.coli_b.subtilis_proms_gc1 %>% as.dendrogram )]
ward_e.coli_b.subtilis_proms_gc1 %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_proms_gc1$labels), col = colvec_gc1) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_proms_gc1$height)*.2) -> to_plot_e.coli_b.subtilis_proms_gc1


#plot(to_plot_e.coli_b.subtilis_proms_gc1, main ='E. coli and B. subtilis promoters: GC')
#legend('top', lty = 1, col = c(e.coli_proms_col, b.subtilis_proms_col), legend = c('Promoters', 'Non-promoters'), bty = 'n')
```

```{r DNA distance promoters both bacteria}
library(ape)
library(dendextend)
#library(fastcluster)

e.coli_b.subtilis_proms <- t(cbind(sapply(e.coli_proms, unlist), sapply(b.subtilis_proms, unlist)))

colvec_e.coli_b.subtilis_proms <- c(rep(e.coli_proms_col, length(e.coli_proms)),
rep('darkgreen', length(b.subtilis_proms)))

#e.coli_proms_non_proms_tv_dist <- dist.dna(as.DNAbin(e.coli_proms_non_proms), model = 'raw', pairwise.deletion = T)

#ward_e.coli_proms_non_proms_tv_dist <- hclust(e.coli_proms_non_proms_tv_dist, method = 'ward.D2')
#or

library(Biostrings)

e.coli_b.subtilis_proms_strings <- apply(e.coli_b.subtilis_proms, 1, function(x) {paste0(x, collapse = '')})
e.coli_b.subtilis_proms_DNAStringSet <- DNAStringSet(e.coli_b.subtilis_proms_strings)

e.coli_b.subtilis_proms_stringdist <- stringDist(e.coli_b.subtilis_proms_DNAStringSet, method = 'levenshtein')

ward_e.coli_b.subtilis_proms_stringdist <- hclust(e.coli_b.subtilis_proms_stringdist, method = 'ward.D2')

colvec_e.coli_b.subtilis <- c(rep(e.coli_proms_col, length(e.coli_proms)),
rep(b.subtilis_proms_col, length(b.subtilis_proms)))
ward_e.coli_b.subtilis_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_proms_stringdist$labels), col = colvec_e.coli_b.subtilis[order.dendrogram(ward_e.coli_b.subtilis_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_proms_stringdist$height)*0.2) -> to_plot_e.coli_b.subtilis_proms_stringdist
#plot(to_plot_e.coli_b.subtilis_proms_stringdist, main = 'E. coli and B. subtilis promoters: DNA Levenshtein distance')

```

```{r DNA distance non_promoters both bacteria}
library(ape)
library(dendextend)
#library(fastcluster)

e.coli_b.subtilis_non_proms <- t(cbind(sapply(e.coli_non_proms, unlist), sapply(b.subtilis_non_proms, unlist)))

colvec_e.coli_b.subtilis_non_proms <- c(rep(e.coli_proms_col, length(e.coli_non_proms)),
rep('blue', length(b.subtilis_non_proms)))

#e.coli_proms_non_proms_tv_dist <- dist.dna(as.DNAbin(e.coli_proms_non_proms), model = 'raw', pairwise.deletion = T)

#ward_e.coli_proms_non_proms_tv_dist <- hclust(e.coli_proms_non_proms_tv_dist, method = 'ward.D2')
#or

library(Biostrings)

#ec.oli
e.coli_proms_non_proms_strings <- sapply(e.coli_proms_non_proms,  function(x) {paste0(x, collapse = '')})
e.coli_proms_non_proms_DNAStringSet <- DNAStringSet(e.coli_proms_non_proms_strings)

e.coli_proms_non_proms_stringdist <- stringDist(e.coli_proms_non_proms_DNAStringSet, method = 'levenshtein')

#b.subtilis
b.subtilis_proms_non_proms_strings <- apply(b.subtilis_proms_non_proms, 1, function(x) {paste0(x, collapse = '')})
b.subtilis_proms_non_proms_DNAStringSet <- DNAStringSet(b.subtilis_proms_non_proms_strings)

b.subtilis_proms_non_proms_stringdist <- stringDist(b.subtilis_proms_non_proms_DNAStringSet, method = 'levenshtein')

#both
e.coli_b.subtilis_non_proms_strings <- apply(e.coli_b.subtilis_non_proms, 1, function(x) {paste0(x, collapse = '')})
e.coli_b.subtilis_non_proms_DNAStringSet <- DNAStringSet(e.coli_b.subtilis_non_proms_strings)

e.coli_b.subtilis_non_proms_stringdist <- stringDist(e.coli_b.subtilis_non_proms_DNAStringSet, method = 'levenshtein')

ward_e.coli_proms_non_proms_stringdist <- hclust(e.coli_proms_non_proms_stringdist, method = 'ward.D2')

ward_e.coli_proms_non_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_stringdist$labels), col = colvec_e.coli_proms_non_proms[order.dendrogram(ward_e.coli_proms_non_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_stringdist$height)*0.2) -> to_plot_e.coli_proms_non_proms_stringdist 


ward_b.subtilis_proms_non_proms_stringdist <- hclust(b.subtilis_proms_non_proms_stringdist, method = 'ward.D2')

ward_b.subtilis_proms_non_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_stringdist$labels), col = colvec_b.subtilis_proms_non_proms[order.dendrogram(ward_b.subtilis_proms_non_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_stringdist$height)*0.2) -> to_plot_b.subtilis_proms_non_proms_stringdist 


ward_b.subtilis_proms_non_proms_stringdist <- hclust(b.subtilis_proms_non_proms_stringdist, method = 'ward.D2')

ward_e.coli_b.subtilis_non_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_non_proms_stringdist$labels), col = colvec_e.coli_b.subtilis_non_proms[order.dendrogram(ward_e.coli_b.subtilis_non_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_non_proms_stringdist$height)*0.2) -> to_plot_e.coli_b.subtilis_non_proms_stringdist 
#plot(to_plot_e.coli_b.subtilis_non_proms_stringdist, main = 'E. coli and B. subtilis non-promoters: DNA Levenshtein distance')

#par(mfrow = c (2,1))
#plot(to_plot_e.coli_b.subtilis_proms_stringdist, main = 'E. coli and B. subtilis promoters: DNA Levenshtein distance')
#plot(to_plot_e.coli_b.subtilis_non_proms_stringdist, main = 'E. coli and B. subtilis non-promoters: DNA Levenshtein distance')
```


```{r assorted plots}
#E01

#par(mfrow = c(4,1))
cex.main = 1.5
plot((to_plot_e.coli_proms_non_proms_E01), main = 'E. coli promoters and non-promoters: E01', cex.main = cex.main)
plot((to_plot_b.subtilis_proms_non_proms_E01), main = 'B. subtilis promoters and non-promoters: E01', cex.main = cex.main)
plot((to_plot_e.coli_b.subtilis_proms_E01), main ='E. coli and B. subtilis promoters: E01', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 

```

```{r }
#par(mfrow = c(4,1))
plot((to_plot_e.coli_proms_non_proms_gc1), main = 'E. coli promoters and non-promoters: GC', cex.main = cex.main)
plot((to_plot_b.subtilis_proms_non_proms_gc1), main = 'B. subtilis promoters and non-promoters: GC', cex.main = cex.main)
plot((to_plot_e.coli_b.subtilis_proms_gc1), main ='E. coli and B. subtilis promoters: GC', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 

```

```{r}
#tanglegram(dend1 = ward_all_4_gc1 %>% as.dendrogram, dend2 = ward_all_4_E01 %>% as.dendrogram)
```


```

```{r EP promoters vs non-promoters for each bacteria}
e.coli_proms_non_proms_mpots <- t(cbind(e.coli_proms_mpots, e.coli_non_proms_mpots))
str(e.coli_proms_non_proms_mpots)

ward_e.coli_proms_non_proms_mpots <- hclust.vector(e.coli_proms_non_proms_mpots, method = 'ward')
plot(ward_e.coli_proms_non_proms_mpots, labels = F, main = 'E.coli promoters and non-promoters (EP)')

colvec_e.coli_mpots <- c(rep(e.coli_proms_col, length(e.coli_proms)),rep(e.coli_non_proms_col, length(e.coli_non_proms)))[order.dendrogram(ward_e.coli_proms_non_proms_mpots %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_e.coli_proms_non_proms_mpots %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_mpots$labels), col = colvec_e.coli_mpots) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_mpots$height)*.2) -> to_plot_e.coli_proms_non_proms_mpots

#plot(to_plot_e.coli_proms_non_proms_mpots)


b.subtilis_proms_non_proms_mpots <- t(cbind(b.subtilis_proms_mpots, b.subtilis_non_proms_mpots))
str(b.subtilis_proms_non_proms_mpots)

ward_b.subtilis_proms_non_proms_mpots <- hclust.vector(b.subtilis_proms_non_proms_mpots, method = 'ward')
plot(ward_b.subtilis_proms_non_proms_mpots, labels = F, main = 'B. subtilis promoters and non-promoters (EP)')

colvec_b.subtilis_mpots <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms)))[order.dendrogram(ward_b.subtilis_proms_non_proms_mpots %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_b.subtilis_proms_non_proms_mpots %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_mpots$labels), col = colvec_b.subtilis_mpots) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_mpots$height)*.2) -> to_plot_b.subtilis_proms_non_proms_mpots

#plot(to_plot_b.subtilis_proms_non_proms_mpots)


e.coli_b.subtilis_proms_mpots <- t(cbind(e.coli_proms_mpots, b.subtilis_proms_mpots))
str(e.coli_b.subtilis_proms_mpots)

ward_e.coli_b.subtilis_proms_mpots<- hclust.vector(e.coli_b.subtilis_proms_mpots, method = 'ward')
plot(ward_b.subtilis_proms_non_proms_mpots, labels = F, main = 'E.coli and B. subtilis promoters (EP)')

colvec_e.coli_b.subtilis_proms_mpots <- c(rep(e.coli_proms_col, length(e.coli_proms)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms)))[order.dendrogram(ward_e.coli_b.subtilis_proms_mpots %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_e.coli_b.subtilis_proms_mpots %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_proms_mpots$labels), col = colvec_e.coli_b.subtilis_proms_mpots) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_proms_mpots$height)*.2) -> to_plot_e.coli_b.subtilis_proms_mpots

#plot(to_plot_e.coli_b.subtilis_proms_mpots)

```

```{r}
##par(mfrow = c(2,2))
cex.main = 1.5
#par(mfrow = c(4,1))
plot(to_plot_e.coli_proms_non_proms_mpots, main ='E. coli promoter and non-promoter mixture (EP)', cex.main = cex.main)
plot(to_plot_b.subtilis_proms_non_proms_mpots, main ='B. subtilis promoter and non-promoter mixture (EP)', cex.main = cex.main)
plot(to_plot_e.coli_b.subtilis_proms_mpots, main ='E. coli and B. subtilis promoter mixture (EP)', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 
```



```{r bendability promoters vs non-promoters for each bacteria}
e.coli_proms_non_proms_bends <- t(cbind(e.coli_proms_bends, e.coli_non_proms_bends))
str(e.coli_proms_non_proms_bends)

ward_e.coli_proms_non_proms_bends <- hclust.vector(e.coli_proms_non_proms_bends, method = 'ward')
plot(ward_e.coli_proms_non_proms_bends, labels = F, main = 'E. coli promoters and non-promoters (bendability)')

colvec_e.coli_bends <- c(rep(e.coli_proms_col, length(e.coli_proms)),rep(e.coli_non_proms_col, length(e.coli_non_proms)))[order.dendrogram(ward_e.coli_proms_non_proms_bends %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_e.coli_proms_non_proms_bends %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_proms_non_proms_bends$labels), col = colvec_e.coli_bends) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_proms_non_proms_bends$height)*.2) -> to_plot_e.coli_proms_non_proms_bends

#plot(to_plot_e.coli_proms_non_proms_bends)


b.subtilis_proms_non_proms_bends <- t(cbind(b.subtilis_proms_bends, b.subtilis_non_proms_bends))
str(b.subtilis_proms_non_proms_bends)

ward_b.subtilis_proms_non_proms_bends <- hclust.vector(b.subtilis_proms_non_proms_bends, method = 'ward')
plot(ward_b.subtilis_proms_non_proms_bends, labels = F, main = 'B. subtilis promoters and non-promoters (bendability)')


colvec_b.subtilis_bends <- c(rep(b.subtilis_proms_col, length(b.subtilis_proms)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms)))[order.dendrogram(ward_b.subtilis_proms_non_proms_bends %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_b.subtilis_proms_non_proms_bends %>% as.dendrogram %>% color_branches(k = length(ward_b.subtilis_proms_non_proms_bends$labels), col = colvec_b.subtilis_bends) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_b.subtilis_proms_non_proms_bends$height)*.2) -> to_plot_b.subtilis_proms_non_proms_bends

#plot(to_plot_b.subtilis_proms_non_proms_bends)


e.coli_b.subtilis_proms_bends <- t(cbind(e.coli_proms_bends, b.subtilis_proms_bends))
str(e.coli_b.subtilis_proms_bends)

ward_e.coli_b.subtilis_proms_bends<- hclust.vector(e.coli_b.subtilis_proms_bends, method = 'ward')
plot(ward_e.coli_b.subtilis_proms_bends, labels = F, main = 'E. coli and B. subtilis promoters (bendability)')


colvec_e.coli_b.subtilis_proms_bends <- c(rep(e.coli_proms_col, length(e.coli_proms)),rep(b.subtilis_non_proms_col, length(b.subtilis_non_proms)))[order.dendrogram(ward_e.coli_b.subtilis_proms_bends %>% as.dendrogram )]

#par(mfrow = c(2,1))
ward_e.coli_b.subtilis_proms_bends %>% as.dendrogram %>% color_branches(k = length(ward_e.coli_b.subtilis_proms_bends$labels), col = colvec_e.coli_b.subtilis_proms_bends) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_e.coli_b.subtilis_proms_bends$height)*.2) -> to_plot_e.coli_b.subtilis_proms_bends

#plot(to_plot_e.coli_b.subtilis_proms_bends)

```

```{r assorted bendability}
#par(mfrow = c(4,1))
cex.main = 1.5
#par(mfrow = c(4,1))
plot(to_plot_e.coli_proms_non_proms_bends, main ='E. coli promoter and non-promoter mixture (bendability)', cex.main = cex.main)
plot(to_plot_b.subtilis_proms_non_proms_bends, main ='B. subtilis promoter and non-promoter mixture (bendability)', cex.main = cex.main)
plot(to_plot_e.coli_b.subtilis_proms_bends, main ='E. coli and B. subtilis promoter mixture (bendability)', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 
```

```{r assorted EP}
#par(mfrow = c(4,1))
cex.main = 1.5
#par(mfrow = c(4,1))
plot(to_plot_e.coli_proms_non_proms_mpots, main ='E. coli promoter and non-promoter mixture (EP)', cex.main = cex.main)
plot(to_plot_b.subtilis_proms_non_proms_mpots, main ='B. subtilis promoter and non-promoter mixture (EP)', cex.main = cex.main)
plot(to_plot_e.coli_b.subtilis_proms_mpots, main ='E. coli and B. subtilis promoter mixture (EP)', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 
```

```{r assorted string distance - Levenshtein}
#par(mfrow = c(4,1))
cex.main = 1.5
#par(mfrow = c(4,1))
plot(to_plot_e.coli_proms_non_proms_stringdist, main ='E. coli promoter and non-promoter mixture (Levenshtein distance)', cex.main = cex.main)
plot(to_plot_b.subtilis_proms_non_proms_stringdist, main ='B. subtilis promoter and non-promoter mixture (Levenshtein distance)', cex.main = cex.main)
plot(to_plot_e.coli_b.subtilis_proms_stringdist, main ='E. coli and B. subtilis promoter mixture (Levenshtein distance)', cex.main = cex.main)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', lty = 1, col = cols_brew, legend = c('E. coli, promoters', 'E.coli, non-promoters', 'B. subtilis, promoters', 'B. subtilis, non-promoters'), bty = 'n', ncol = 1, cex = 3, lwd = 3)
 
```


```{r substring occurences}

library(ngram)
library(Biostrings)


#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(e.coli_proms, e.coli_non_proms, b.subtilis_proms, b.subtilis_non_proms)){
  print(length(i))
  tmp <- DNAString(paste0(unlist(i), collapse = ''))
  
  #tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}



colnames(all_seqs_oligos_freq) <- c('e.coli_proms', 'e.coli_non_proms', 'b.subtilis_proms', 'b.subtilis_non_proms')
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]


par(mfrow = c(3,2))
ylim = c (0, .005)

for (i in 1:ncol(all_seqs_oligos_freq)){
  barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

```

```{r hexamers occurences PCA, pca 1-2}
PCA12 <- prcomp(all_seqs_oligos_freq[,c(1,2)], scale = T)
library(factoextra)
fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')
```

```{r hexamers occurences PCA, pca 3-4}
PCA34 <- prcomp(all_seqs_oligos_freq[,c(3,4)], scale = T)
fviz_pca_biplot(PCA34, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')
```

```{r hexamers occurences PCA, pca 1-3}
PCA13 <- prcomp(all_seqs_oligos_freq[,c(1,3)], scale = T)
fviz_pca_biplot(PCA13, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')
```

```{r hexamers occurences PCA, pca 1-4}
PCA13 <- prcomp(all_seqs_oligos_freq[,c(1:4)], scale = T)
fviz_pca_biplot(PCA13, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')

```

```{r E. coli promoters melting for different sigma values - genes}
library(clusterProfiler)
load('/home/jane/Загрузки/mydf_genes.rda')

mydf_genes_split <- split(mydf_genes, mydf_genes$group)
lapply(mydf_genes_split, nrow)  
sapply(mydf_genes_split, function(x) {paste0(as.character(x$Entrez), collapse = ', ')})
#compareCluster(Entrez~group, mydf_genes[1:148,], fun = 'enrichGO', OrgDb = "org.EcK12.eg.db")


```

```{r GO MF overrepresentaion test by panther. Dumb way}
#diff_sigmas_panther_overrepresentaion_test <- list()
for (i in levels(mydf_genes$group)) {
   tmp <- read.table(paste0('/home/jane/Документы/Misha/mol_a_2018/panther_', i, '_mf_stat_test.txt'), header = T, skip =11, sep = '\t')
  assign(paste0(i, '_panther_overrepresentaion_test'), tmp)

}

diff_sigmas_panther_overrepresentaion_test <- grep('*_panther_overrepresentaion_test', ls(), value = T)


GO.molecular.function.complete <- sapply(diff_sigmas_panther_overrepresentaion_test, FUN = function(x){(get(x))$GO.molecular.function.complete})

Client.Text.Box.Input..expected. <- sapply(diff_sigmas_panther_overrepresentaion_test, FUN = function(x){(get(x))$Client.Text.Box.Input..expected})


#which mf made it to the top, max nrow in 2 member of the list
which.max(sapply(GO.molecular.function.complete, length))
GO.molecular.function.complete[[2]]

par(mfrow = c(1,4),
las=1, mar=c(10,0.5,0.5, 0.5))
ylim = c(0, 125)
for (i in seq_along(GO.molecular.function.complete)){
barplot(Client.Text.Box.Input..expected.[[i]], names.arg =  substr(as.character(GO.molecular.function.complete[[i]]), 0, 15), las = 2, ylim = ylim)
}

all_levels <- levels(sapply(GO.molecular.function.complete, function(x) {x[[1]]}))

new_df <- matrix(NA, ncol = length(GO.molecular.function.complete), nrow = length(all_levels))
rownames(new_df) <- all_levels
colnames(new_df) <- names(GO.molecular.function.complete)
for (i in seq_along(GO.molecular.function.complete)){
  nos <- which(all_levels %in% levels(GO.molecular.function.complete[[i]]) )
  new_df[nos,i] <- Client.Text.Box.Input..expected.[[i]][nos]
}


View(new_df[1:16,]) #therest are empty cases

par(mfrow = c(1,2))
matplot(t(new_df[1:16,]), type = 'h', col = rainbow(16),  lty = 1:16, lwd = 4)
plot(1, type="n", axes=F, xlab="", ylab="")
legend('center', col = rainbow(16),  legend = substr(rownames(new_df[1:16,]),0,50), cex = .6, ncol = 1, lwd = 3,lty = 1:16)


barplot(c(Client.Text.Box.Input..expected.[[1]], Client.Text.Box.Input..expected.[[2]]), names.arg =  substr(as.character(GO.molecular.function.complete[[1]]), 0, 15), las = 2, ylim = ylim)

library(tidyr)
df %>% spread(f2, v1, fill=0) %>% gather(f2, v1, -f1)
```

