---
title: "Mycoplasma_vs_Synechocystis_vs_E.coli"
author: "Mikhail Orlov"
date: '6 ноября 2017 г '
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Clearing workspace, setting working directory and loading R packages
```{r}
rm(list=ls())

#setwd('/home/mikhail/Documents/Script_2016_all/PCA/')

library(R.matlab)
library(seqinr)
library(factoextra)
#library(data.table)
library(caret)
#library(doMC)
library(Biostrings)
library(ape)
library(reldna)
library(dendextend)
```
Reading in E. coli K12 genome (GenBank Accesion U00096.2), creating reverse complement genome and tranformation genome into character form 
```{r Reading in E. coli genome}
e.coli_U00096.2<-paste0(read.GenBank(access.nb = 'U00096.2', as.character = T)$U00096.2, collapse = '')
reverseComplement_e.coli_U00096.2<-as.character(reverseComplement(DNAString(e.coli_U00096.2)))

e.coli_U00096.2_char<-unlist(strsplit(e.coli_U00096.2, ''))
reverseComplement_e.coli_U00096.2_char<-unlist(strsplit(reverseComplement_e.coli_U00096.2, ''))

```

Reading in Synechocystis sp. PCC6803 complete , creating reverse complement genome and tranformation genome into character form 
```{r Reading in Synechocystis sp. PCC6803 complete genome, in supplementary noted genbank acession NC_000911.1, warning=FALSE, message=FALSE, cache=TRUE}
synechocystis_NC_000911.1 <- toupper(paste0(read.GenBank(access.nb = 'NC_000911.1', as.character = T)$NC_000911.1, collapse = ''))
reverseComplement_synechocystis_NC_000911.1 <- as.character(reverseComplement(DNAString(synechocystis_NC_000911.1)))

synechocystis_NC_000911.1_char<-unlist(strsplit(synechocystis_NC_000911.1, ''))
reverseComplement_synechocystis_NC_000911.1<-unlist(strsplit(reverseComplement_synechocystis_NC_000911.1, ''))

```

Reading in M. gallisepticum strain S6 genome (CP006916.3 assembly), creating reverse complement genome and tranformation genome into character form 
```{r Reading in M. gallisepticum strain S6 genome (CP006916.3 assembly), warning=FALSE, message=FALSE, cache=TRUE}

mycoplasma_CP006916.3<- toupper(paste0(read.GenBank(access.nb = 'CP006916.3', as.character = T)$CP006916.3, collapse = ''))
reverseComplement_mycoplasms_CP006916.3<- as.character(reverseComplement(DNAString(mycoplasma_CP006916.3)))

mycoplasma_CP006916.3_char<-unlist(strsplit(mycoplasma_CP006916.3, ''))
reverseComplement_mycoplasma_CP006916.3<-unlist(strsplit(mycoplasma_CP006916.3, ''))

```



Dynamical properties calculation from within R using octave script translated according to Grinevich et al., 2013
Creating function for dynamical properties calculation

```{r dynchars function }



dynchars<-function(seq, interval_size) {
  if (missing(seq))
    stop("Need to specify sequence (as a vector of chars)")
  
  if (missing(interval_size))
    stop("Need to specify interval size")
  
  if(!is.character(seq))
    stop("Sequence must be a character vector containing A, C, G, T letters only")
  
seq<-toupper(seq)
seq<-c(seq, seq[2:(interval_size)])
  
a<-3.4*10^(-10)
I<-c(7.6, 4.8, 8.2, 4.1)*10^(-44)
K<-c(227, 155, 220, 149)*10^(-20)
V<-c(2.09, 1.43, 3.12, 2.12)*10^(-20)
tau<-c(127, 99, 140, 84)
  
csA<-cumsum(seq=='A') 
csT<-cumsum(seq=='T')
csG<-cumsum(seq=='G')
csC<-cumsum(seq=='C')

countA = csA[interval_size:length(csA)]-c(0, csA[1:(length(csA)-interval_size)])
countT = csT[interval_size:length(csT)]-c(0, csT[1:(length(csT)-interval_size)])
countG = csG[interval_size:length(csG)]-c(0, csG[1:(length(csG)-interval_size)])
countC = csC[interval_size:length(csC)]-c(0, csC[1:(length(csC)-interval_size)])
  
M<-cbind(countA, countT, countG, countC)/interval_size
M_comp<-cbind(countT, countA, countC, countG)/interval_size
M_comp<-apply(t(M_comp),1,rev) 
Is<-as.numeric(M%*%I)#! numeric conversion
Ks<-as.numeric(M%*%K)
Vs<-as.numeric(M%*%V)
  
E01<-(8*(Ks*Vs)^0.5)* 6E23 / 4184
d1<-((Ks*a^2)/Vs)^(0.5)/a;
c1<-(Ks*a^2/Is)^0.5
m1<-E01/c1/6.011E-26
taus1<-as.numeric(M%*%tau) #!as.numeric conversion
gc1 = M[,3] + M[,4]
  
Is<-as.numeric(M_comp%*%I)#! numeric conversion
Ks<-as.numeric(M_comp%*%K)
Vs<-as.numeric(M_comp%*%V)

  
E02<- 8*(Ks*Vs)^0.5  * 6E23 / 4184;
d2<-((Ks*a^2)/Vs)^(0.5)/a;
c2<-(Ks*a^2/Is)^0.5;
m2<-E02/c2/6.011E-26;
taus2<-as.numeric(M_comp%*%tau)
gc2 = M_comp[,3] + M_comp[,4]

dynchars_return<-list(E01=E01, d1=d1, c1=c1, m1=m1, taus1=taus1, gc1=gc1, E02=E02, d2=d2, c2=c2, m2=m2, taus2=taus2, gc2=gc2)
  
return(dynchars_return)
  
}
```

First to set interval size for dynchars
```{r Genome-wide characteristics profiels for three bacteria, warning=FALSE, message=FALSE, cache=TRUE}

interval_size <- 100
whole_e.coli_dynchars <- dynchars(e.coli_U00096.2_char, interval_size = interval_size)
whole_e.coli_EP <- lseqspline1D(e.coli_U00096.2, bound = c(1, nchar(e.coli_U00096.2)), width = 1, ref =  1)

whole_synechocystis_dynchars <- dynchars(synechocystis_NC_000911.1_char, interval_size = interval_size)
whole_synechocystis_EP <- lseqspline1D(synechocystis_NC_000911.1, bound = c(1, nchar(synechocystis_NC_000911.1)), width = 1, ref =  1)

whole_mycoplasma_dynchars<- dynchars(mycoplasma_CP006916.3_char, interval_size = interval_size)
whole_mycoplasma_EP <- lseqspline1D(mycoplasma_CP006916.3, bound = c(1, nchar(mycoplasma_CP006916.3)), width = 1, ref =  1)

par(mfrow = c(3,4))
rbind(sapply(whole_e.coli_dynchars, mean),
sapply(whole_synechocystis_dynchars, mean),
sapply(whole_mycoplasma_dynchars, mean))


par(mfrow = c (1,3))
ylim = c(-0.07, -0.017)

boxplot(whole_e.coli_EP$mpot, ylim = ylim)
boxplot(whole_synechocystis_EP$mpot, ylim = ylim, main = 'EP for 3 bacteria')
boxplot(whole_mycoplasma_EP$mpot, ylim = ylim)


```


Loading data on sequences of different types (promoters, non-promoters, genes, islands, and lowscore) from .Rdata files (must be copied separetely)
```{r}


load('/home/jane/Документы/Misha/Lab/mol_a-16/spline_dataset_pro.Rdata')

```


Check-up for reverse strand
```{r, echo=FALSE, warning=FALSE, message=FALSE, cache=TRUE}
gregexpr(toupper(dataset_pro$aaeRp10$seq), reverseComplement_e.coli_U00096.2)
nchar(e.coli_U00096.2)-dataset_pro$aaeRp10$tss-270
```
Extracting data on all promoters and on experimentaly found ones - including previosely calculated electrostatic potential profiles
```{r}
pro_names<-names(dataset_pro)
tsss<-c()
seqs<-c()
exp_strands<-c()
exp_tsss<-c()
exp_EP_check<-c() #for experimentally found one
exp_names<-c()

for (i in 1:length(dataset_pro)){
  tsss<-c(tsss, dataset_pro[[i]]$tss)
  seqs<-c(seqs, dataset_pro[[i]]$seq)
  if (dataset_pro[[i]]$evidence=='experimental'){ 
    #exp_proms<-rbind(exp_proms, c(pro_names[i], strands[i], tsss[i], seqs[i]))
    exp_names<-c(exp_names, pro_names[i])
    exp_EP_check<-rbind(exp_EP_check, dataset_pro[[i]]$mpot)  	
    #exp_promoters<-rbind(exp_promoters, c(pro_names[i], strands[i], tsss[i], seqs[i]))
    exp_tsss<-c(exp_tsss, dataset_pro[[i]]$tss) 
    exp_strands<-c(exp_strands, dataset_pro[[i]]$strand)}
}
```

# # # Changing genomic coordinate used for dataset_... - TSS is according to 5'-end for both strands 
```{r}

exp_tsss[which(exp_strands=='reverse')]<-nchar(e.coli_U00096.2)-exp_tsss[which(exp_strands=='reverse')]
#exp_tsss<--exp_tsss
```


```{r strReverse funtcion}
## a useful function: rev() for strings
strReverse <- function(x)
        sapply(lapply(strsplit(x, NULL), rev), paste, collapse="")
```


Creating matrices for data on activation energy ('E01') and size ('d1'). The matrices is 699*201 - 699 promoters sequences, 201 value for a physical property profile
to 5'-end for both strands 


Creating matrices of dynamical properties and GC-content for intervals [-150; 50] nts
```{r E. coli promoters physics, warning=FALSE, message=FALSE, cache=TRUE}
for (i in c('seqsforward', 'seqsreverse', 'E01forward', 'E01reverse', 'd1forward', 'd1reverse', 'GCforward', 'GCreverse', 'EPforward', 'EPreverse')) {
  assign(i, c())
}
zout<- -480:239


for (i in 1:length(exp_tsss)) {
  if (exp_strands[i]=='forward') {
     seqsforward<-c(seqsforward, substr(e.coli_U00096.2, start = exp_tsss[i] - 150, stop = exp_tsss[i] + 50))
                                    
    E01forward<-rbind(E01forward, whole_e.coli_dynchars$E01[as.numeric(exp_tsss[i]-150):(exp_tsss[i]+50)])
    #aeos2forward<-rbind(aeos2forward, matr1[(as.numericexp_tsss[i]-150):(as.numericexp_tsss[i]+50),2])
    d1forward<-rbind(d1forward, whole_e.coli_dynchars$d1[(as.numeric(exp_tsss[i])-150):(exp_tsss[i]+50)])
    #aeos4forward<-rbind(aeos4forward, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    GCforward<-rbind(GCforward, whole_e.coli_dynchars$gc1[(as.numeric(exp_tsss[i])-150):(as.numeric(exp_tsss[i])+50)])
    
    p<-lseqspline1D(substr(e.coli_U00096.2, exp_tsss[i]-250, exp_tsss[i]+150), bound=c(50, 350), ref=251 )
    EPforward<-rbind(EPforward, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
     } else {
    seqsreverse<-c(seqsreverse, substr(e.coli_U00096.2, start = exp_tsss[i] - 150, stop = exp_tsss[i] + 50)) #string itself is revesed, so there is no need to mirror intervals
    E01reverse<-rbind(E01reverse, whole_e.coli_dynchars$E02[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)])
    #aeos2reverse<-rbind(aeos2reverse, matr1[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150),2])
    d1reverse<-rbind(d1reverse, whole_e.coli_dynchars$d2[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)])
    #aeos4reverse<-rbind(aeos4reverse, matr1[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150),4])
    GCreverse<-rbind(GCreverse, (whole_e.coli_dynchars$gc2[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)]))
    
    
    
    p<-lseqspline1D(substr(reverseComplement_e.coli_U00096.2, exp_tsss[i]-250, exp_tsss[i]+150), bound=c(50, 350), ref=251 )
    EPreverse<-rbind(EPreverse, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
    
    }
}


#merging matrices for forward and reverse strands  together
seqs <- toupper(c(seqsforward, sapply(seqsreverse, FUN = strReverse)))
 E01<-rbind( E01forward, t(apply( E01reverse, 1, rev)))
#aeos2<-rbind(aeos2forward, aeos2reverse)
 d1<-rbind( d1forward, t(apply( d1reverse, 1, rev)))
#aeos4<-rbind(aeos4forward, aeos4reverse)
 GC<-rbind( GCforward, t(apply( GCreverse, 1, rev)))

 EP<-rbind( EPforward, t(apply( EPreverse, 1, rev)))
```




```{r Datasets from An experimentally anchored map of transcriptional start sites in the model cyanobacterium Synechocystis sp. PCC6803 Jan Mitschke, 2011}

synechocystis_proms <- (read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/synechocystis_proms_datasets_1015154108_sd01.csv', skip = 14, header = T))
```



Creating matrices of dynamical properties and GC-content for intervals [-150; 50] nts
```{r Synechocystis proms physics, warning=FALSE, message=FALSE, cache=TRUE}


for (i in c('synechocystis_seqsforward', 'synechocystis_seqsreverse', 'synechocystis_E01forward', 'synechocystis_E01reverse', 'synechocystis_d1forward', 'synechocystis_d1reverse', 'synechocystis_GCforward', 'synechocystis_GCreverse', 'synechocystis_EPforward', 'synechocystis_EPreverse')) {
  assign(i, c())
}
zout<- -480:239
interval_size <- 50
for (i in 1:length(synechocystis_proms$X1.TSS)) {
  if (synechocystis_proms$X4.strand[i]=='fwd') {
    
    synechocystis_seqsforward<-c(synechocystis_seqsforward, substr(synechocystis_NC_000911.1, start = synechocystis_proms$X1.TSS[i] - 150, stop = synechocystis_proms$X1.TSS[i] + 50))
                                    
                                    
    synechocystis_E01forward<-rbind(synechocystis_E01forward,
dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$E01)                       
    synechocystis_d1forward<-rbind(synechocystis_d1forward, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$d1)                       
    #aeos4forward<-rbind(aeos4forward, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    synechocystis_GCforward<-rbind(synechocystis_GCforward, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(synechocystis_NC_000911.1, synechocystis_proms$X1.TSS[i]-250, synechocystis_proms$X1.TSS[i]+150), bound=c(50, 350), ref=251 )
    synechocystis_EPforward<-rbind(synechocystis_EPforward, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
     } else  {
    synechocystis_seqsreverse<-c(synechocystis_seqsreverse, substr(synechocystis_NC_000911.1, start = synechocystis_proms$X1.TSS[i] - 50, stop = synechocystis_proms$X1.TSS[i] + 150))
       
    synechocystis_E01reverse<-rbind(synechocystis_E01reverse,
dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$E01)                       
    synechocystis_d1reverse<-rbind(synechocystis_d1reverse, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$d1)                       
    #aeos4reverse<-rbind(aeos4reverse, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    synechocystis_GCreverse<-rbind(synechocystis_GCreverse, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(synechocystis_NC_000911.1, synechocystis_proms$X1.TSS[i]-150, synechocystis_proms$X1.TSS[i]+250), bound=c(50, 350), ref=251 )
    synechocystis_EPreverse<-rbind(synechocystis_EPreverse, 
                        ##mpot_whole_genome_reverse$mpot[mpot_whole_genome_reverse$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
    
    
    
    }
}


#merging matrices for forward and reverse strands  together
synechocystis_seqs <- toupper(c(synechocystis_seqsforward, sapply(synechocystis_seqsreverse, FUN = strReverse)))
synechocystis_E01<-rbind(synechocystis_E01forward, t(apply(synechocystis_E01reverse, 1, rev)))
#aeos2<-rbind(aeos2forward, aeos2reverse)
synechocystis_d1<-rbind(synechocystis_d1forward, t(apply(synechocystis_d1reverse, 1, rev)))
#aeos4<-rbind(aeos4forward, aeos4reverse)
synechocystis_GC<-rbind(synechocystis_GCforward, t(apply(synechocystis_GCreverse, 1, rev)))

synechocystis_EP<-rbind(synechocystis_EPforward, t(apply(synechocystis_EPreverse, 1, rev)))

```




```{r  Reading in M. gallisepticum strain S6 promoters from Fisunov et al.}

mycoplasma_proms <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/mycoplasma_gallisepticum_fisunov_2016.csv', header = T)

plot(seq_along(mycoplasma_CP006916.3_char), type = 'n', main = 'Mycoplasma gallisepticum promoters location', xlab = 'Sequence (nts)', yaxt = 'n')
abline(v = mycoplasma_proms$Position, col = 'red')

#table(mycoplasma_proms$Strand)
```

Right flank needs to be extended by 250 nts
```{r Mycoplasma extension}
extended_by <- 250
extended_mycoplasma_CP006916.3 <- paste0(mycoplasma_CP006916.3,
                                         substr(mycoplasma_CP006916.3, nchar(mycoplasma_CP006916.3)-extended_by, nchar(mycoplasma_CP006916.3)))

extended_mycoplasma_CP006916.3_char <- strsplit(extended_mycoplasma_CP006916.3, split = '')[[1]]
```

Creating matrices of dynamical properties and GC-content for intervals [-150; 50] nts
```{r mycoplasma proms physics, warning=FALSE, message=FALSE, cache=TRUE}


for (i in c('mycoplasma_seqsforward', 'mycoplasma_seqsreverse', 'mycoplasma_E01forward', 'mycoplasma_E01reverse', 'mycoplasma_d1forward', 'mycoplasma_d1reverse', 'mycoplasma_GCforward', 'mycoplasma_GCreverse', 'mycoplasma_EPforward', 'mycoplasma_EPreverse')) {
  assign(i, c())
}
zout<- -480:239
interval_size <- 50
for (i in 1:length(mycoplasma_proms$Position)) {
  if (mycoplasma_proms$Strand[i]=='fwd') {
    
    mycoplasma_seqsforward<-c(mycoplasma_seqsforward, substr(extended_mycoplasma_CP006916.3, start = mycoplasma_proms$Position[i] - 150, stop = mycoplasma_proms$Position[i] + 50))
                                    
                                    
    mycoplasma_E01forward<-rbind(mycoplasma_E01forward,
dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-150):(mycoplasma_proms$Position[i]+50)], interval_size = interval_size)$E01)                       
    mycoplasma_d1forward<-rbind(mycoplasma_d1forward, dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-150):(mycoplasma_proms$Position[i]+50)], interval_size = interval_size)$d1)                       
    #aeos4forward<-rbind(aeos4forward, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    mycoplasma_GCforward<-rbind(mycoplasma_GCforward, dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-150):(mycoplasma_proms$Position[i]+50)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(extended_mycoplasma_CP006916.3, mycoplasma_proms$Position[i]-250, mycoplasma_proms$Position[i]+150), bound=c(50, 350), ref=251 )
    mycoplasma_EPforward<-rbind(mycoplasma_EPforward, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
     } else  {
    mycoplasma_seqsreverse<-c(mycoplasma_seqsreverse, substr(extended_mycoplasma_CP006916.3, start = mycoplasma_proms$Position[i] - 50, stop = mycoplasma_proms$Position[i] + 150))
       
    mycoplasma_E01reverse<-rbind(mycoplasma_E01reverse,
dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-50):(mycoplasma_proms$Position[i]+150)], interval_size = interval_size)$E01)                       
    mycoplasma_d1reverse<-rbind(mycoplasma_d1reverse, dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-50):(mycoplasma_proms$Position[i]+150)], interval_size = interval_size)$d1)                       
    #aeos4reverse<-rbind(aeos4reverse, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    mycoplasma_GCreverse<-rbind(mycoplasma_GCreverse, dynchars(extended_mycoplasma_CP006916.3_char[(mycoplasma_proms$Position[i]-50):(mycoplasma_proms$Position[i]+150)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(extended_mycoplasma_CP006916.3, mycoplasma_proms$Position[i]-150, mycoplasma_proms$Position[i]+250), bound=c(50, 350), ref=251 )
    mycoplasma_EPreverse<-rbind(mycoplasma_EPreverse, 
                        ##mpot_whole_genome_reverse$mpot[mpot_whole_genome_reverse$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
    
    
    
    }
}


#merging matrices for forward and reverse strands  together
mycoplasma_seqs <- toupper(c(mycoplasma_seqsforward, sapply(mycoplasma_seqsreverse, FUN = strReverse)))
mycoplasma_E01<-rbind(mycoplasma_E01forward, t(apply(mycoplasma_E01reverse, 1, rev)))
#aeos2<-rbind(aeos2forward, aeos2reverse)
mycoplasma_d1<-rbind(mycoplasma_d1forward, t(apply(mycoplasma_d1reverse, 1, rev)))
#aeos4<-rbind(aeos4forward, aeos4reverse)
mycoplasma_GC<-rbind(mycoplasma_GCforward, t(apply(mycoplasma_GCreverse, 1, rev)))

mycoplasma_EP<-rbind(mycoplasma_EPforward, t(apply(mycoplasma_EPreverse, 1, rev)))

```


```{r Plotting characteristics profiles for 3 bacteria, eval=F}

par(mfrow = c(4,3),
    mar = rep(0.5,4))

matplot(E01, type = 'l', col = 1)
matplot(d1, type = 'l', col = 1)
matplot(GC, type = 'l', col = 1)
matplot(EP, type = 'l', col = 1)

matplot(synechocystis_E01, type = 'l', col = 1)
matplot(synechocystis_d1, type = 'l', col = 1)
matplot(synechocystis_GC, type = 'l', col = 1)
matplot(synechocystis_EP, type = 'l', col = 1)

matplot(mycoplasma_E01, type = 'l', col = 1)
matplot(mycoplasma_d1, type = 'l', col = 1)
matplot(mycoplasma_GC, type = 'l', col = 1)
matplot(mycoplasma_EP, type = 'l', col = 1)

```


```{r string distance, warning=FALSE, message=FALSE, cache=TRUE}
library(Biostrings)

mycoplasma_synechocystis_e.coli_proms_strings <- sapply(c(mycoplasma_seqs, synechocystis_seqs, seqs),  function(x) {paste0(x, collapse = '')})
mycoplasma_synechocystis_e.coli_proms_DNAStringSet <- DNAStringSet(mycoplasma_synechocystis_e.coli_proms_strings)

mycoplasma_synechocystis_e.coli_proms_stringdist<- stringDist(mycoplasma_synechocystis_e.coli_proms_DNAStringSet, method = 'levenshtein')

ward_mycoplasma_synechocystis_e.coli_proms_stringdist<- hclust(mycoplasma_synechocystis_e.coli_proms_stringdist, method = 'ward.D2')

colvec_mycoplasma_synechocystis_e.coli_proms<- c(rep('magenta', length(mycoplasma_proms$Position)), rep('darkgreen', length(synechocystis_proms$X1.TSS)),rep('orange', length(exp_tsss)))

ward_mycoplasma_synechocystis_e.coli_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_mycoplasma_synechocystis_e.coli_proms_stringdist$order), col = colvec_mycoplasma_synechocystis_e.coli_proms[order.dendrogram(ward_mycoplasma_synechocystis_e.coli_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_mycoplasma_synechocystis_e.coli_proms_stringdist$height)*0.2) -> to_plot_ward_mycoplasma_synechocystis_e.coli_proms_stringdist


# #plot(to_plot_ward_mycoplasma_synechocystis_e.coli_proms_stringdist, main = 'E. coli, Synechocystis sp., and Mycoplasma gallisepticum promoters:\n DNA Levenshtein distance')
```



```{r three bacteria promoters mixture - all the characteristics, warning=FALSE, message=FALSE, cache=TRUE, fig.width = 10, fig.asp = 1}
library(fastcluster)
library(dendextend)
cex.main = 2


par(mfrow = c(5,1),
    mar = rep(1.2,4),
    oma = rep(1.2,4))

to_plot_wards <- list()
for (i in c('E01', 'd1', 'GC', 'EP')){
  three_tmp <- rbind(get(i), get(paste0('synechocystis_', i)), get(paste0('mycoplasma_', i)))
  ward_three_tmp<- hclust.vector(t(scale(t(three_tmp), center = T, scale = T)), method = 'ward') #!!

  colvec_ward_three_tmp <- c(rep('orange', length(exp_tsss)), rep('darkgreen',length(synechocystis_proms$X1.TSS)), rep('magenta',length(mycoplasma_proms$Position)))[order.dendrogram(ward_three_tmp %>% as.dendrogram )]

  ward_three_tmp %>% as.dendrogram %>% color_branches(k = length(ward_three_tmp$order), col = colvec_ward_three_tmp) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_three_tmp$height)*.2) -> to_plot_ward_three_tmp

  plot(to_plot_ward_three_tmp, labels = F, main = paste0('E.coli, Synechocystis sp., and Mycoplasma gallisepticum promoters (', i, ')'), cex.main = cex.main )
  
  assign(paste0('to_plot_wards', i), ward_three_tmp)
}

plot(to_plot_ward_mycoplasma_synechocystis_e.coli_proms_stringdist, main = 'E. coli and Synechocystis promoters: DNA Levenshtein distance', cex.main = cex.main )

legend('topright', legend = c('E.coli', 'Synechocystis sp.', 'Mycoplasma gallisepticum'), fill = rev(unique(colvec_mycoplasma_synechocystis_e.coli_proms)))
```



```{r substring occurences, warning=FALSE, message=FALSE, cache=TRUE, fig.width = 10, fig.asp = 1}

library(Biostrings)


#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(seqs, synechocystis_seqs, mycoplasma_seqs, e.coli_U00096.2, synechocystis_NC_000911.1, mycoplasma_CP006916.3)){
  ###print(length(i))
  tmp <- DNAString(paste0(unlist(i), collapse = ''))
  
  #tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}



colnames(all_seqs_oligos_freq) <- c('E. coli promoters', 'Synechocystis sp. promoters', 'Mycoplasma gallisepticum promoters', 'E. coli\n complete genome', 'Synechocystis sp.\n complete genome', 'Mycoplasma gallisepticum\n complete genome')
# #View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]


par(mfrow = c(2,3))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
  barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

```


```{r hexamers occurences PCA, pca 1-2, warning=FALSE, message=FALSE, cache=TRUE}
PCA <- prcomp(all_seqs_oligos_freq, scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')

label.size <- 1.7
#library(ggfortify)
#autoplot(PCA,
    #     shape = FALSE,
   #      loadings = TRUE,
  #       label = TRUE,
 #        loadings.label = TRUE,
#         label.size = label.size)+
 # theme_bw()

library(factoextra)

fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=0.8)
  
```

## hexanucleotides vs their frequences
```{r Other text representation, warning=FALSE, message=FALSE, cache=TRUE, results = 'hide'}

#panel.cor function
panel.cor <- function(x, y, digits=2, prefix="", cex.cor) 
{
    usr <- par("usr"); on.exit(par(usr)) 
    par(usr = c(0, 1, 0, 1)) 
    r <- abs(cor(x, y)) 
    txt <- format(c(r, 0.123456789), digits=digits)[1] 
    txt <- paste(prefix, txt, sep="") 
    if(missing(cex.cor)) cex <- 0.8/strwidth(txt) 
 
    test <- cor.test(x,y) 
    # borrowed from printCoefmat
    Signif <- symnum(test$p.value, corr = FALSE, na = FALSE, 
                  cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                  symbols = c("***", "**", "*", ".", " ")) 
 
    text(0.5, 0.5, txt, cex = cex * r) 
    text(.8, .8, Signif, cex=cex, col=2) 
}

pairs(all_seqs_oligos_freq, lower.panel=panel.smooth, upper.panel=panel.cor)

```


```{r Oligonucleotide occurences, warning=FALSE, message=FALSE, cache=TRUE, fig.width = 10, fig.asp = 1}

#par(mfrow = c(2,4))
#apply(all_seqs_oligos_freq, 2, function(x) {
 # boxplot(x, ylim = range(all_seqs_oligos_freq))
#  }
 # )

for (i in 1:ncol(all_seqs_oligos_freq)){
  assign(paste0('lm', i), lm(all_seqs_oligos_freq[,i] ~ seq_along(all_seqs_oligos_freq[,i])))
}
```

```{r Descriptive statistics for hexanucleotide frequences, warning=FALSE, message=FALSE, cache=TRUE}

par(mfrow = c(6,5),
    mar = rep(2,4))
plotnames <- c('E. coli promoters', 'Synechocystis sp. promoters', 'Mycoplasma gallisepticum promoters', 'E. coli complete genome', 'Synechocystis sp. complete genome', 'Mycoplasma gallisepticum complete genome')
for (i in 1:ncol(all_seqs_oligos_freq)){
 
  plot(1, type="n", axes=F, xlab="", ylab="")
  mtext(plotnames[i], side = 3, col = 2, cex= 0.7, line = -3)

  plot(get(paste0('lm', i)))
 
}
#boxplot(all_seqs_oligos_freq[,1])
#identify(rep(1, length(all_seqs_oligos_freq[,1])), all_seqs_oligos_freq[,1], labels = #seq_along(all_seqs_oligos_freq[,1]))
```


```{r cuplcl usage}
#source('/home/mikhail/Documents/Script_2016_all/cut_and_plot_clusters.R')

#function to cut heirarchical clusterization results at given clusters numer and plot all pfrofiles from each

cuplcl <- function(data, method, k, main) {
  library(fastcluster)
  library(dendextend)
  hclust <- hclust.vector(data, method = method)
  cut_hclust<-cutree(hclust, k=k)
  m <- rbind(rep(1, k), c(2:(k+1)))
  layout(m)
  par(mar = c(1.5,1.5, 1, 1),
      oma = c(2,2,2,2))

  dend <- color_branches(as.dendrogram(hclust), col = cut_hclust[order.dendrogram(as.dendrogram(hclust))])
plot(dend%>%set('labels_cex', NA), main = main, cex.main = 1.75)
  
  ylim <- range(data)
  for (i in unique(cut_hclust[order.dendrogram(as.dendrogram(hclust))])) {
    matplot(t(data[which(cut_hclust==i),]), type='l', col=i, ylim = ylim)
 # #   print(c(i, names(which(cut_hclust==i))))
  }
}
```





#Separate characteristics for separate bacteria clusterization
```{r cluster analysis for various properties and 3 bacteria on E01, warning=FALSE, message=FALSE, cache=TRUE}
library(fastcluster)
ward_ecoli_E01 <- hclust.vector(E01, method = 'ward')
ward_synechocystis_E01 <- hclust.vector(synechocystis_E01, method = 'ward')
ward_mycoplasma_E01 <- hclust.vector(mycoplasma_E01, method = 'ward')

#plot(ward_ecoli_E01, labels = F)
#plot(ward_synechocystis_E01, labels = F)
#plot(ward_mycoplasma_E01, labels = F)

cuplcl(E01, method = 'ward', k = 4, main = 'E. coli, E01')
cuplcl(synechocystis_E01, method = 'ward', k = 4, main = 'Synechocystis, E01')
cuplcl(mycoplasma_E01, method = 'ward', k = 4, main = 'Mycoplasma gallisepticum, E01')

```

```{r an interesting cluster for Mycoplasma E01, warning=FALSE, message=FALSE, cache=TRUE}

par(mfrow = c(2,1),
    mar = c(1.5, 1, 1.5,1),
    oma = rep(1.5,4))
`Sequence (nts)` <- -150:50
inds <- which(cutree(ward_mycoplasma_E01, k = 3) == 3)
matplot(`Sequence (nts)`, t(mycoplasma_E01[inds,]), type = 'l', col = 1, main = 'Mycoplasma gallisepticum: small (13 promoters) cluster E01 profiles', ylab = 'E01 value' )
abline(v = 0, lty = 3)

pos <- mycoplasma_proms[inds,'Position']
##mycoplasma_proms[inds,]

#the promoters genomic location
interv <- seq_along(mycoplasma_CP006916.3_char)
plot(interv, type = 'n', ylab = '', xlab = 'Sequence (nts)', yaxt = 'n', xlim = range(interv))
abline(v = pos , col = 'darkred')


#the interseting small cluster sequence

##mycoplasma_seqs[inds]
library(ggseqlogo)
library(ggplot2)

#logo for [-50;50] interval
mycoplasma_peculiar_E01cluster_short_seqs <- sapply(mycoplasma_seqs[inds], function(x) {substr(x, start = 100, stop = 200)})
ggseqlogo(mycoplasma_peculiar_E01cluster_short_seqs,  ncol = 1)+ scale_y_continuous(name="Информационнное содержание (бит)")+scale_x_discrete(name ="Последовательность относительно ТСТ (п.н.)", limits=1:100, labels = -50:50)+coord_fixed(ratio = 2)
#does the cluster representatives differ from the universe?


#logo for [-150;50] interval

ggseqlogo(mycoplasma_seqs[inds],  ncol = 1)+ scale_y_continuous(name="Информационнное содержание (бит)")+
scale_x_discrete(name ="Последовательность относительно ТСТ (п.н.)", limits=1:200, labels = -150:50)+coord_fixed(ratio = 2)
#does the cluster representatives differ from the universe?

#ape sweater plot

library(ape)

mycoplasma_peculiar_E01cluster_short_seqs_chars <- lapply(mycoplasma_peculiar_E01cluster_short_seqs, FUN = function(x) {strsplit(x, split = '')[[1]]})
# #image.DNAbin(as.DNAbin(mycoplasma_peculiar_E01cluster_short_seqs_chars))

#string bidtance for the cluster
library(Biostrings)
range(stringDist(x = mycoplasma_peculiar_E01cluster_short_seqs, method = 'lev'))

plot(hclust(stringDist(x = mycoplasma_peculiar_E01cluster_short_seqs, method = 'lev')), labels = F)
```

```{r Same for scaled versions E01, warning=FALSE, message=FALSE, cache=TRUE}
cuplcl(t(scale(t(E01), center = T, scale = T)), method = 'ward', main = 'Scaled E01, E. coli', k = 4)
cuplcl(t(scale(t(synechocystis_E01), center = T, scale = T)), method = 'ward', main = 'Scaled E01, Synechocystis', k = 4)
cuplcl(t(scale(t(mycoplasma_E01), center = T, scale = T)), method = 'ward', main = 'Scaled E01, Mycoplasma gallisepticum', k = 4)

```

```{r cluster analysis for various properties and 3 bacteria on EP, warning=FALSE, message=FALSE, cache=TRUE} 
library(fastcluster)
ward_ecoli_EP <- hclust.vector(EP, method = 'ward')
ward_synechocystis_EP <- hclust.vector(synechocystis_EP, method = 'ward')
ward_mycoplasma_EP <- hclust.vector(mycoplasma_EP, method = 'ward')

#plot(ward_ecoli_EP, labels = F)
#plot(ward_synechocystis_EP, labels = F)
#plot(ward_mycoplasma_EP, labels = F)

cuplcl(EP, method = 'ward', k = 4, main = 'E. coli, EP')
cuplcl(synechocystis_EP, method = 'ward', k = 4, main = 'Synechocystis, EP')
cuplcl(mycoplasma_EP, method = 'ward', k = 4, main = 'Mycoplasma gallisepticum, EP')
```

```{r another interesting small cluster for Mycoplasma EP, warning=FALSE, message=FALSE, cache=TRUE}

par(mfrow = c(2,1),
    mar = c(1.5, 1, 1.5,1),
    oma = rep(1.5,4))
`Sequence (angstrom)` <- -480:239
inds <- which(cutree(ward_mycoplasma_EP, k = 3) == 3)
matplot(`Sequence (angstrom)`, t(mycoplasma_EP[inds,]), type = 'l', col = 1, main = 'Mycoplasma gallisepticum: small (13 promoters) cluster EP profiles', ylab = 'EP value' )
abline(v = 0, lty = 3)

pos <- mycoplasma_proms[inds,'Position']
##mycoplasma_proms[inds,]
#the promoters genomic location
interv <- seq_along(mycoplasma_CP006916.3_char)
plot(interv, type = 'n', ylab = '', xlab = 'Sequence (nts)', yaxt = 'n', xlim = range(interv))
abline(v = pos , col = 'darkred')


#the interseting small cluster sequence

##mycoplasma_seqs[inds]
library(ggseqlogo)
library(ggplot2)

#logo for [-50;50] interval
mycoplasma_peculiar_EPcluster_short_seqs <- sapply(mycoplasma_seqs[inds], function(x) {substr(x, start = 100, stop = 200)})
ggseqlogo(mycoplasma_peculiar_EPcluster_short_seqs,  ncol = 1)+ scale_y_continuous(name="Информационнное содержание (бит)")+
scale_x_discrete(name ="Последовательность относительно ТСТ (п.н.)", limits=1:100, labels = -50:50)+coord_fixed(ratio = 2)
#does the cluster representatives differ from the universe?


#logo for [-150;50] interval

ggseqlogo(mycoplasma_seqs[inds],  ncol = 1)+ scale_y_continuous(name="Информационнное содержание (бит)")+
scale_x_discrete(name ="Последовательность относительно ТСТ (п.н.)", limits=1:200, labels = -150:50)+coord_fixed(ratio = 2)
#does the cluster representatives differ from the universe?

#string bidtance for the cluster
library(Biostrings)
range(stringDist(x = mycoplasma_peculiar_EPcluster_short_seqs, method = 'lev'))
#3 peculiar cluster promoters and their relation to heatshock expression from FIsunov et al.

mycoplasma_peculiar_EPcluster_short_seqs_chars <- lapply(mycoplasma_peculiar_EPcluster_short_seqs, FUN = function(x) {strsplit(x, split = '')[[1]]})
#small intersting clusters for Mycoplsma - on E01 and EP

par(mfrow = c(2,1))
image.DNAbin(as.DNAbin(mycoplasma_peculiar_EPcluster_short_seqs_chars))
image.DNAbin(as.DNAbin(mycoplasma_peculiar_E01cluster_short_seqs_chars))



par(mfrow = c(2,1),
    mar = rep(1.25, 4))
boxplot(as.numeric(gsub(as.vector(mycoplasma_proms[inds,'Control1']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[inds,'Control2']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[inds,'Heatshock1']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[inds,'Heatshock2']), pattern = ',', replacement = '.')))

boxplot(as.numeric(gsub(as.vector(mycoplasma_proms[,'Control1']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[,'Control2']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[,'Heatshock1']), pattern = ',', replacement = '.')),
as.numeric(gsub(as.vector(mycoplasma_proms[,'Heatshock2']), pattern = ',', replacement = '.')))


#NB  Papazisi et al. (2003) reported that the predicted oriC region of M. gallisepticum was located between the dnaN and dnaA genes. The predicted oriC region contains five DnaA box sequences, with the consensus sequence 5′-TTWTMHAMA-3′... 180 nts long

#from genebank by assecion :complement(984237..985412) /gene="dnaN"
#gene 3161..4546 /gene="dnaA" /locus_tag="GCW_00010"
dnaA <- 3161:4546
```

```{r Same for scaled versions EP, warning=FALSE, message=FALSE, cache=TRUE}
cuplcl(t(scale(t(EP), center = T, scale = T)), method = 'ward', main = 'Scaled E01, E. coli', k = 4)
cuplcl(t(scale(t(synechocystis_EP), center = T, scale = T)), method = 'ward', main = 'Scaled E01, Synechocystis sp.', k = 4)
cuplcl(t(scale(t(mycoplasma_EP), center = T, scale = T)), method = 'ward', main = 'Scaled E01, Mycoplasma gallisepticum', k = 4)

```



```{r cluster analysis for various properties and 3 bacteria on GC, warning=FALSE, message=FALSE, cache=TRUE}
library(fastcluster)
ward_ecoli_GC <- hclust.vector(GC, method = 'ward')
ward_synechocystis_GC <- hclust.vector(synechocystis_GC, method = 'ward')
ward_mycoplasma_GC <- hclust.vector(mycoplasma_GC, method = 'ward')

#plot(ward_ecoli_GC, labels = F)
#plot(ward_synechocystis_GC, labels = F)
#plot(ward_mycoplasma_GC, labels = F)

cuplcl(t(GC), method = 'ward', k = 4, main = 'E. coli, GC')
cuplcl(t(synechocystis_GC), method = 'ward', k = 4, main = 'Synechocystis, GC')
cuplcl(t(mycoplasma_GC), method = 'ward', k = 4, main = 'Mycoplasma gallisepticum, GC')

```

```{r Same for scaled versions GC, warning=FALSE, message=FALSE, cache=TRUE}
cuplcl(t(scale(t(GC), center = T, scale = T)), method = 'ward', main = 'Scaled GC, E. coli', k = 4)
cuplcl(t(scale(t(synechocystis_GC), center = T, scale = T)), method = 'ward', main = 'Scaled GC, Synechocystis', k = 4)
cuplcl(t(scale(t(mycoplasma_GC), center = T, scale = T)), method = 'ward', main = 'Scaled GC, Mycoplasma gallisepticum', k = 4)

```


```{r Reading in transription units data, warning=FALSE, message=FALSE, cache=TRUE}

TUSet <- read.table('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/regulondb_TUSet_html_line_removed.txt', header = T, sep = '\t')
```

Promoter numbers and genenames
```{r E01 clusters population and GOstats, warning=FALSE, message=FALSE, cache=TRUE}

library(org.EcK12.eg.db)
cut_ward_ecoli_E01 <- cutree(ward_ecoli_E01, k = 4)

#table(cut_ward_ecoli_E01)

ecoli_E01_promoters_clusters <- c()
for (i in unique(cut_ward_ecoli_E01)) {
 assign(paste0('ecoli_E01_clust_', i),
       exp_names[which(cut_ward_ecoli_E01 ==i)])
  
  #alternatively
  tmp <- exp_names[which(cut_ward_ecoli_E01 ==i)]
  tmp1 <- cbind(rep(i, length(tmp)), tmp)
  
  ecoli_E01_promoters_clusters <- rbind(ecoli_E01_promoters_clusters, tmp1)
}


#promoters to genes (complete set)

genes_from_promoters_list <- c()
for (i in seq_along(exp_names)){
  tmp <- as.character(TUSet$Name.of.the.gene.s..contained.in.the.transcription.unit[which(TUSet$Promoter.Name == exp_names[i])])
  if(length(tmp) ==0) {tmp <- NA}
  tmp <- paste0(tmp, collapse = ',')
  genes_from_promoters_list <- c(genes_from_promoters_list, tmp)
}


#all the genes universe without repeating and 2 or more names in 1 vector member

fixed_vectors_genes_from_promoters_list <- strsplit(paste0(genes_from_promoters_list, collapse = ','), split = ',')[[1]]
fixed_vectors_genes_from_promoters_list <-fixed_vectors_genes_from_promoters_list[(fixed_vectors_genes_from_promoters_list)!='NA']

#clusters of genes as fixed vectors
for (i in unique(cut_ward_ecoli_E01)){
  genes_clust_tmp <- strsplit(paste0(genes_from_promoters_list[which(cut_ward_ecoli_E01 == i)], collapse = ','), split = ',')[[1]]
  if ('NA' %in% genes_clust_tmp) {genes_clust_tmp <- genes_clust_tmp[-which(genes_clust_tmp =='NA')]}
  assign(paste0('genes_clust', i), unique(genes_clust_tmp))
}


## Bimap interface:
# Convert the object to a list
xx <- as.list(org.EcK12.egALIAS2EG)
# Remove pathway identifiers that do not map to any entrez gene id
xx <- xx[!is.na(xx)]

#common names in universe (xx) and the samples

setdiffed <- setdiff(fixed_vectors_genes_from_promoters_list, names(xx))
fixed_vectors_genes_from_promoters_list <- fixed_vectors_genes_from_promoters_list[!fixed_vectors_genes_from_promoters_list%in%setdiffed]
entrezIds <- mget(fixed_vectors_genes_from_promoters_list, envir=org.EcK12.egALIAS2EG)
  ###   haveEntrezId <- names(entrezIds)[sapply(entrezIds, function(x) !is.na(x))]

#common names in universe (xx) and the clusters 1-4

setdiffed_clust1 <- setdiff(genes_clust1, names(xx))
fixed_genes_clust1 <- genes_clust1[!genes_clust1%in%setdiffed_clust1]
entrezIds_clust1 <- mget(fixed_genes_clust1, envir=org.EcK12.egALIAS2EG)


setdiffed_clust2 <- setdiff(genes_clust2, names(xx))
fixed_genes_clust2 <- genes_clust2[!genes_clust2%in%setdiffed_clust2]
entrezIds_clust2 <- mget(fixed_genes_clust2, envir=org.EcK12.egALIAS2EG)


setdiffed_clust3 <- setdiff(genes_clust3, names(xx))
fixed_genes_clust3 <- genes_clust3[!genes_clust3%in%setdiffed_clust3]
entrezIds_clust3 <- mget(fixed_genes_clust3, envir=org.EcK12.egALIAS2EG)

setdiffed_clust4 <- setdiff(genes_clust4, names(xx))
fixed_genes_clust4 <- genes_clust4[!genes_clust4%in%setdiffed_clust4]
entrezIds_clust4 <- mget(fixed_genes_clust4, envir=org.EcK12.egALIAS2EG)

dfGO_ecoli_E01 <- as.data.frame(rbind(cbind(rep(1, length(entrezIds_clust1)), as.character(entrezIds_clust1)),
      cbind(rep(2, length(entrezIds_clust2)), as.character(entrezIds_clust2)),
      cbind(rep(3, length(entrezIds_clust3)), as.character(entrezIds_clust3)),
      cbind(rep(4, length(entrezIds_clust4)), as.character(entrezIds_clust4))
))

colnames(dfGO_ecoli_E01) <- c('group', 'Entrezid')
  
# #save(dfGO_ecoli_E01, file = '/home/jane/Документы/Misha/mol_a_2018/dfGO_ecoli_E01.rda')
###   haveEntrezId <- names(entrezIds)[sapply(entrezIds, function(x) !is.na(x))]
```




```{r GO statistics using GOstats for E01, warning=FALSE, message=FALSE, cache=TRUE}
library(GOstats)
library(Category)
library(DBI)
library(org.EcK12.eg.db)

load('/home/jane/Документы/Misha/mol_a_2018/dfGO_ecoli_E01.rda')

entrezUniverse <- as.character(dfGO_ecoli_E01$Entrezid)

selectedEntrezIds_clust1 <- as.character(dfGO_ecoli_E01$Entrezid[which(dfGO_ecoli_E01$group == 1)])
selectedEntrezIds_clust2 <- as.character(dfGO_ecoli_E01$Entrezid[which(dfGO_ecoli_E01$group == 2)])
selectedEntrezIds_clust3 <- as.character(dfGO_ecoli_E01$Entrezid[which(dfGO_ecoli_E01$group == 3)])
selectedEntrezIds_clust4 <- as.character(dfGO_ecoli_E01$Entrezid[which(dfGO_ecoli_E01$group == 4)])

#setting cutoff value
hgCutoff <- 0.0005
#hgCutoff <- 0.1

#selecting onthology

ontology <- 'BP'
#ontology <- 'MF'

for (i in 1:4) {
params <- new("GOHyperGParams",
 geneIds=get(paste0('selectedEntrezIds_clust', i)),
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

  hgOver_clust <- hyperGTest(params)
  
  assign(x = paste0('df_summary_clust', i), summary(hgOver_clust))
}


head(df_summary_clust1) #primarly nucleoside metabolisim and ion transport
#View(df_summary_clust2) #suflur,bicarbonic acid, aldehyde metabolism
#View(df_summary_clust3) #biogenesis, aerobic respiration
#View(df_summary_clust4) #sugar metabolism
#We would also like to perform a conditional test.  Instead of having to de ne a new
#GO-HyperGParamsinstance  by  hand,  we  can  create  a  copy  and  update  just  the  parameter  of interest.
#paramsCond <- params
#conditional(paramsCond) <- TRUE
#hgCondOver <- hyperGTest(paramsCond)


#hgOver_clust1 <- hyperGTest(params_clust1)
#hgOver_clust2 <- hyperGTest(params_clust2)
#hgOver_clust3 <- hyperGTest(params_clust3)
#hgOver_clust4 <- hyperGTest(params_clust4)


#df_clust1 <- summary(hgOver_clust1)
#df_clust2 <- summary(hgOver_clust2)
#df_clust3 <- summary(hgOver_clust3)
#df_clust4 <- summary(hgOver_clust4)

########cbind(rbind(df_clust2, df_clust3, df_clust4), (df_clust2, df_clust3, df_clust4)

#names(df_clust2)

```



```{r cluster analysis for various properties and 3 bacteria on d1, warning=FALSE, message=FALSE, cache=TRUE}
library(fastcluster)
ward_ecoli_d1 <- hclust.vector(d1, method = 'ward')
ward_synechocystis_d1 <- hclust.vector(synechocystis_d1, method = 'ward')
ward_mycoplasma_d1 <- hclust.vector(mycoplasma_d1, method = 'ward')

par(mfrow = c (1,3))

#plot(ward_ecoli_d1, labels = F)
#plot(ward_synechocystis_d1, labels = F)
#plot(ward_mycoplasma_d1, labels = F)

cuplcl(t(d1), method = 'ward', k = 4, main = 'E. coli, d1')
cuplcl(t(synechocystis_d1), method = 'ward', k = 4, main = 'Synechocystis, d1')
cuplcl(t(mycoplasma_d1), method = 'ward', k = 4, main = 'Mycoplasma gallisepticum, d1')

```

Promoter numbers and genenames
```{r d1 clusters population and GOstats, warning=FALSE, message=FALSE, cache=TRUE}
cut_ward_ecoli_d1 <- cutree(ward_ecoli_d1, k = 4)

#table(cut_ward_ecoli_d1)

ecoli_d1_promoters_clusters <- c()
for (i in unique(cut_ward_ecoli_d1)) {
 assign(paste0('ecoli_d1_clust_', i),
       exp_names[which(cut_ward_ecoli_d1 ==i)])
  
  #alternatively
  tmp <- exp_names[which(cut_ward_ecoli_d1 ==i)]
  tmp1 <- cbind(rep(i, length(tmp)), tmp)
  
  ecoli_d1_promoters_clusters <- rbind(ecoli_d1_promoters_clusters, tmp1)
}


#promoters to genes (complete set)

genes_from_promoters_list <- c()
for (i in seq_along(exp_names)){
  tmp <- as.character(TUSet$Name.of.the.gene.s..contained.in.the.transcription.unit[which(TUSet$Promoter.Name == exp_names[i])])
  if(length(tmp) ==0) {tmp <- NA}
  tmp <- paste0(tmp, collapse = ',')
  genes_from_promoters_list <- c(genes_from_promoters_list, tmp)
}


#all the genes universe without repeating and 2 jr more names in 1 vector member

fixed_vectors_genes_from_promoters_list <- strsplit(paste0(genes_from_promoters_list, collapse = ','), split = ',')[[1]]
fixed_vectors_genes_from_promoters_list <-fixed_vectors_genes_from_promoters_list[(fixed_vectors_genes_from_promoters_list)!='NA']

#clusters of genes as fixed vectors
for (i in unique(cut_ward_ecoli_d1)){
  genes_clust_tmp <- strsplit(paste0(genes_from_promoters_list[which(cut_ward_ecoli_d1 == i)], collapse = ','), split = ',')[[1]]
  if ('NA' %in% genes_clust_tmp) {genes_clust_tmp <- genes_clust_tmp[-which(genes_clust_tmp =='NA')]}
  assign(paste0('genes_clust', i), unique(genes_clust_tmp))
}


## Bimap interface:
# Convert the object to a list
xx <- as.list(org.EcK12.egALIAS2EG)
# Remove pathway identifiers that do not map to any entrez gene id
xx <- xx[!is.na(xx)]

#common names in universe (xx) and the samples

setdiffed <- setdiff(fixed_vectors_genes_from_promoters_list, names(xx))
fixed_vectors_genes_from_promoters_list <- fixed_vectors_genes_from_promoters_list[!fixed_vectors_genes_from_promoters_list%in%setdiffed]
entrezIds <- mget(fixed_vectors_genes_from_promoters_list, envir=org.EcK12.egALIAS2EG)
  ###   haveEntrezId <- names(entrezIds)[sapply(entrezIds, function(x) !is.na(x))]

#common names in universe (xx) and the clusters 1-4

setdiffed_clust1 <- setdiff(genes_clust1, names(xx))
fixed_genes_clust1 <- genes_clust1[!genes_clust1%in%setdiffed_clust1]
entrezIds_clust1 <- mget(fixed_genes_clust1, envir=org.EcK12.egALIAS2EG)


setdiffed_clust2 <- setdiff(genes_clust2, names(xx))
fixed_genes_clust2 <- genes_clust2[!genes_clust2%in%setdiffed_clust2]
entrezIds_clust2 <- mget(fixed_genes_clust2, envir=org.EcK12.egALIAS2EG)


setdiffed_clust3 <- setdiff(genes_clust3, names(xx))
fixed_genes_clust3 <- genes_clust3[!genes_clust3%in%setdiffed_clust3]
entrezIds_clust3 <- mget(fixed_genes_clust3, envir=org.EcK12.egALIAS2EG)

setdiffed_clust4 <- setdiff(genes_clust4, names(xx))
fixed_genes_clust4 <- genes_clust4[!genes_clust4%in%setdiffed_clust4]
entrezIds_clust4 <- mget(fixed_genes_clust4, envir=org.EcK12.egALIAS2EG)

dfGO_ecoli_d1 <- as.data.frame(rbind(cbind(rep(1, length(entrezIds_clust1)), as.character(entrezIds_clust1)),
      cbind(rep(2, length(entrezIds_clust2)), as.character(entrezIds_clust2)),
      cbind(rep(3, length(entrezIds_clust3)), as.character(entrezIds_clust3)),
      cbind(rep(4, length(entrezIds_clust4)), as.character(entrezIds_clust4))
))

colnames(dfGO_ecoli_d1) <- c('group', 'Entrezid')
  
# #save(dfGO_ecoli_d1, file = '/home/jane/Документы/Misha/mol_a_2018/dfGO_ecoli_d1.rda')
###   haveEntrezId <- names(entrezIds)[sapply(entrezIds, function(x) !is.na(x))]
```




```{r GO statistics using GOstats for d1, warning=FALSE, message=FALSE, cache=TRUE}
library(GOstats)
library(Category)
library(DBI)
library(org.EcK12.eg.db)

load('/home/jane/Документы/Misha/mol_a_2018/dfGO_ecoli_d1.rda')

entrezUniverse <- as.character(dfGO_ecoli_d1$Entrezid)

selectedEntrezIds_clust1 <- as.character(dfGO_ecoli_d1$Entrezid[which(dfGO_ecoli_d1$group == 1)])
selectedEntrezIds_clust2 <- as.character(dfGO_ecoli_d1$Entrezid[which(dfGO_ecoli_d1$group == 2)])
selectedEntrezIds_clust3 <- as.character(dfGO_ecoli_d1$Entrezid[which(dfGO_ecoli_d1$group == 3)])
selectedEntrezIds_clust4 <- as.character(dfGO_ecoli_d1$Entrezid[which(dfGO_ecoli_d1$group == 4)])

#setting cutoff value
hgCutoff <- 0.0005
#hgCutoff <- 0.1

#selecting onthology

ontology <- 'BP'
#ontology <- 'MF'

for (i in 1:4) {
params <- new("GOHyperGParams",
 geneIds=get(paste0('selectedEntrezIds_clust', i)),
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

  hgOver_clust <- hyperGTest(params)
  
  assign(x = paste0('df_summary_clust', i), summary(hgOver_clust))
}


#View(df_summary_clust1) #-
#View(df_summary_clust2) #ion transport
#View(df_summary_clust3) # citrate cycle, monosacharide
# #View(df_summary_clust4) #lipid metabolism
#We would also like to perform a conditional test.  Instead of having to de ne a new
#GO-HyperGParamsinstance  by  hand,  we  can  create  a  copy  and  update  just  the  parameter  of interest.
#paramsCond <- params
#conditional(paramsCond) <- TRUE
#hgCondOver <- hyperGTest(paramsCond)


#hgOver_clust1 <- hyperGTest(params_clust1)
#hgOver_clust2 <- hyperGTest(params_clust2)
#hgOver_clust3 <- hyperGTest(params_clust3)
#hgOver_clust4 <- hyperGTest(params_clust4)


#df_clust1 <- summary(hgOver_clust1)
#df_clust2 <- summary(hgOver_clust2)
#df_clust3 <- summary(hgOver_clust3)
#df_clust4 <- summary(hgOver_clust4)

########cbind(rbind(df_clust2, df_clust3, df_clust4), (df_clust2, df_clust3, df_clust4)

#names(df_clust2)

```


```{r Data for GO, eval=FALSE }

library(org.EcK12.eg.db)

#not found in regulon TUset
 # absent_genenames <- c('cpxQ', 'insD', 'efeU_1', "efeU_2", "sroD", 'tcyJ', "gadF", "istR-1","istR-2", "phnE_1", "phnE_2", "dgcT", "ilvG_1", "ilvG_2"  )

  
Entrezid <- c()
group <- c()

LIST <- list(genes_clust1, genes_clust2, genes_clust3, genes_clust4)
names(LIST) <- c('cl1', 'cl2', 'cl3', 'cl4')
for ( i in  seq_along(LIST)) {
#  if ('TRUE' %in% unique(absent_genenames %in% LIST[[i]])) {
  #  tmp <- LIST[[i]][-which(LIST[[i]] %in% absent_genenames)]
 # }
  
  #handling absent genes
  for (gene in LIST[[i]]){
     get(x = gene, envir=org.EcK12.eg.db::org.EcK12.egALIAS2EG)
  }
  mget(x= as.character(LIST[[i]]), envir=org.EcK12.eg.db::org.EcK12.egALIAS2EG)
  Entrezid <- c(Entrezid, unlist(mget(x= as.character(tmp), envir=org.EcK12.eg.db::org.EcK12.egALIAS2EG)))
  group <- c(group,  rep(i, length(unlist(mget(x= as.character(tmp), envir=org.EcK12.eg.db::org.EcK12.egALIAS2EG)))))
 # dfGO <- rbind(dfGO, cbind(Entrezid, group))
}

dfGO <- data.frame(Entrezid = Entrezid, group = group)
                   
                   
#colnames(dfGO) <- c('Entrezids', 'group')
save(dfGO, file = '/home/jane/Документы/Misha/mol_a_2018/different_bacteria_physics_an_text/ecoli_sidd_clustering_df_to_go.rda')


genes_formula <- compareCluster(Entrezid~group, data=dfGO, fun='groupGO', OrgDb='org.EcK12.eg.db')
genes_formula@compareClusterResult

```

loading previously saved SIDD profiles
```{r SIDD clusterization, warning=FALSE, message=FALSE, cache=TRUE}

sidds_ecoli <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_ecoli.csv')[,-1] #rownumbers in column1

sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis.csv')[,-1] #rownumbers in column1



library(fastcluster)
synechocystis_e.coli_proms_sidds <- cbind(sidds_ecoli, sidds_syn)


ward_synechocystis_e.coli_proms_sidd <- hclust.vector(t(synechocystis_e.coli_proms_sidds), method = 'ward')

#plot(ward_synechocystis_e.coli_proms_sidd, labels = F)

colvec_ward_synechocystis_e.coli_proms_sidd <- c(rep('darkgreen', length(synechocystis_proms$X1.TSS)),
rep('orange', length(exp_tsss)))
ward_synechocystis_e.coli_proms_sidd %>% as.dendrogram %>% color_branches(k = length(ward_synechocystis_e.coli_proms_sidd$order), col = colvec_ward_synechocystis_e.coli_proms_sidd[order.dendrogram(ward_synechocystis_e.coli_proms_sidd %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_synechocystis_e.coli_proms_sidd$height)*0.2) -> to_plot_ward_synechocystis_e.coli_proms_sidd #circlize_dendrogram(labels=F, dend_track_height = 0.95) 
  

plot(to_plot_ward_synechocystis_e.coli_proms_sidd, labels = F, main = 'E. coli and Synechocystis sp. promoters, SIDD profiles')

```


```{r #tanglegrams3, cache=T}

#labels(to_plot_wards[[1]]) <- labels(to_plot_wards[[2]]) <- as.character(labels(to_plot_wards[[2]]))


#for (i in c('E01', 'd1', 'GC', 'EP')){
#for (j in c('E01', 'd1', 'GC', 'EP')){

 #   if (i == j) {next} else {
      
#tanglegram(to_plot_wardsE01, to_plot_wardsd1, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsE01, to_plot_wardsGC, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsE01, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)


#tanglegram(to_plot_wardsd1, to_plot_wardsGC, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsd1, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)

#tanglegram(to_plot_wardsGC, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)
  #plot(to_plot_ward_both_tmp, labels = F, main = paste0('E.coli and Synechocystis promoters (', i, ')'), cex.main = cex.main )
  #  }
  #}
#}
```


```{r E. coli 1500 nts long sidd, warning=FALSE, message=FALSE, cache=TRUE}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_ecoli,ty='l',lty =1, col=1, lwd =0.5, main = 'SIDD profiles, E. coli')

ward_ecoli_1500sidd<- hclust.vector(t(sidds_ecoli), method = 'ward')
#plot(ward_ecoli_1500sidd, labels = F)
cut_ecoli_1500sidd <- cutree(ward_ecoli_1500sidd, k = 4)

cuplcl(data = t(sidds_ecoli), method = 'ward', main = 'SIDD E. coli', k = 4)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```


```{r Synechocystis 1500 nts long sidd, warning=FALSE, message=FALSE, cache=TRUE}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_syn,ty='l',lty =1, col=1, lwd =0.5, main = 'SIDD, Synechocystis sp.')

ward_synechocystis_1500sidd<- hclust.vector(t(sidds_syn), method = 'ward')
#plot(ward_synechocystis_1500sidd, labels = F)
cut_synechocystis_1500sidd <- cutree(ward_synechocystis_1500sidd, k = 4)

cuplcl(data = t(sidds_syn), method = 'ward', main = 'SIDD Synechocystis sp.', k = 4)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```



```{r Whole-genome SIDD vectors, warning=FALSE, message=FALSE, cache=TRUE, eval = FALSE}
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_2500_500.rda')
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_5000_1000.rda')
load('/home/jane/Документы/Misha/Lab/mol_a-16/spline_dataset_pro.Rdata')

exp_pro_numbers<- which(sapply(dataset_pro, function(x) {x$evidence =='experimental' })==T)
exp_pro_tsss <- sapply(dataset_pro, function(x){x$tss})[exp_pro_numbers]


par(mfrow = c(2,1))
interv <- 1:100000
plot(whole_e.coli_sidd_2500_500[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )
plot(whole_e.coli_sidd_5000_1000[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )

exp_pro_sidd_2500_500 <- sapply(exp_pro_tsss, function(x) {whole_e.coli_sidd_2500_500[(x-1000):(x+1000)]})
matplot(exp_pro_sidd_2500_500, type = 'l', col =1, lwd =0.3)

exp_pro_sidd_5000_1000 <- sapply(exp_pro_tsss, function(x) {whole_e.coli_sidd_5000_1000[(x-1000):(x+1000)]})
matplot(exp_pro_sidd_5000_1000, type = 'l', col =1, lwd =0.3)
```


```{r Rollaply max, warning=FALSE, message=FALSE, cache=TRUE, eval = FALSE}
library(zoo)

rollmax_whole_e.coli_sidd_2500_500 <- rollmax(x = whole_e.coli_sidd_2500_500, k = 250)

rollmax_whole_e.coli_sidd_5000_1000 <- rollmax(x = whole_e.coli_sidd_5000_1000, k = 250)


par(mfrow = c(2,1),
    mar = rep(0.7,4))
interv <- 1:50000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )
plot(rollmax_whole_e.coli_sidd_5000_1000[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )

which_rollmax_whole_e.coli_sidd_2500_500_0.95 <- which(rollmax_whole_e.coli_sidd_2500_500>0.95)
which_rollmax_whole_e.coli_sidd_5000_1000_0.95 <- which(rollmax_whole_e.coli_sidd_5000_1000>0.95)                                              
#how big is a moltem part of the genome (nts)
length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)
length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)

par(mfrow = c(2,1),
    mar = rep(0.7,4))
interv <- 1:500000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_2500_500_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)), pch = '|', col = 2)
abline(v = exp_pro_tsss, col = 'darkgreen' )

plot(rollmax_whole_e.coli_sidd_5000_1000[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_5000_1000_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)), pch = '|', col = 2)
abline(v = exp_pro_tsss, col = 'darkgreen' )

#which regions are melted for different split sets
interv <- 300000:400000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_2500_500_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)), pch = '|', col = 2)
text(x = 0, y =0.95, labels = '2500-500\n fragments', cex = 0.7)
points(x = which_rollmax_whole_e.coli_sidd_5000_1000_0.95, y = rep(0.75, length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)), pch = '|', col = 2)
text(x = 0, y =0.75, labels = '5000-1000\n fragments', cex = 0.7)
points(x = exp_pro_tsss, y = rep(0.55, length(exp_pro_tsss)), pch = '|', col = 'darkblue')
text(x = 0, y =0.55, labels = 'Experimentally\n evidenced\n promoters', cex = 0.7)

```
Column 1 contains numbers and is removed 
```{r CHecking bimodality for SIDD for all promoters, warning=FALSE, message=FALSE, cache=TRUE, eval = FALSE}

interv <- 750:2250
interv <- 500:2500
####interv <- 900:2100
####interv <- 1000:2000

sums_sidds_ecoli <- apply(sidds_ecoli[,-1], 2, function(x) {sum(x[interv])})

sums_sidds_syn <- apply(sidds_syn[,-1], 2, function(x) {sum(x[interv])})


#svg('/home/jane/Документы/Misha/mol_a_2018/different_bacteria_physics_an_text/densities_auc_sidd_2_bacteria.svg', height = 10, width = 10)
par(mfrow = c(2,1))
hist(sums_sidds_ecoli, breaks = 15, freq = F, xlab = 'AUC for 1500 nts-long SIDD profile', main = 'Histogram of E. coli Promoters AUC SIDD Data\n with Kernel Density Plot')
lines(density(sums_sidds_ecoli, na.rm = T))


hist(sums_sidds_syn, breaks = 15, freq = F, xlab = 'AUC for 1500 nts-long SIDD profile', main = 'Histogram of Synechocystis sp. Promoters AUC SIDD Data\n with Kernel Density Plot')
lines(density(sums_sidds_syn, na.rm = T))

#dev.off()

library(diptest)
diptest_sums_sidds_ecoli <- dip.test(sums_sidds_ecoli, simulate.p.value = T, B = 20000) #D = 0.055838, p-value < 2.2e-16 for 1500 nts -long interval
# D = 0.029814, p-value = 1e-04 for 2000 nts
diptest_sums_sidds_ecoli

diptest_sums_sidds_syn<- dip.test(sums_sidds_syn, simulate.p.value = T, B = 20000) #D = 0.037924, p-value < 2.2e-16 for 1500 nts -long interval
#D = 0.045919, p-value < 2.2e-16 for 2000 nts
diptest_sums_sidds_syn
```


```{r if dynamic properties ditributed unimodally, eval=FALSE}


funcs <- c('sum', 'mean', 'median', 'max')

for (i in funcs){
  assign(paste0(i,' _E01_e.coli_promoters'), apply(E01, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_E01_e.coli_promoters'))))
}


for (i in funcs){
  assign(paste0(i,'_d1_e.coli_promoters'), apply(d1, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_d1_e.coli_promoters'))))
}


for (i in funcs){
  assign(paste0(i,'_GC_e.coli_promoters'), apply(GC, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_GC_e.coli_promoters'))))
}

for (i in funcs){
  assign(paste0(i,'_EP_e.coli_promoters'), apply(GC, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_GC_e.coli_promoters'))))
}


```

```{r GO statistics using GOstats, warning=FALSE, message=FALSE, cache=TRUE}
library(GOstats)
library(Category)
library(DBI)
library(org.EcK12.eg.db)

load('/home/jane/Документы/Misha/mol_a_2018/different_bacteria_physics_an_text/ecoli_sidd_clustering_df_to_go.rda')

entrezUniverse <- as.character(dfGO$Entrezid)
selectedEntrezIds_clust2 <- as.character(dfGO$Entrezid[which(dfGO$group == 2)])
selectedEntrezIds_clust3 <- as.character(dfGO$Entrezid[which(dfGO$group == 3)])
selectedEntrezIds_clust4 <- as.character(dfGO$Entrezid[which(dfGO$group == 4)])

#setting cutoff value
hgCutoff <- 0.001
#hgCutoff <- 0.1

#selecting onthology

ontology <- 'BP'
#ontology <- 'MF'

params_clust2 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust2,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

params_clust3 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust3,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

params_clust4 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust4,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

#We would also like to perform a conditional test.  Instead of having to de ne a new
#GO-HyperGParamsinstance  by  hand,  we  can  create  a  copy  and  update  just  the  parameter  of interest.
#paramsCond <- params
#conditional(paramsCond) <- TRUE
#hgCondOver <- hyperGTest(paramsCond)



hgOver_clust2 <- hyperGTest(params_clust2)
hgOver_clust3 <- hyperGTest(params_clust3)
hgOver_clust4 <- hyperGTest(params_clust4)



df_clust2 <- summary(hgOver_clust2)
df_clust3 <- summary(hgOver_clust3)
df_clust4 <- summary(hgOver_clust4)

########cbind(rbind(df_clust2, df_clust3, df_clust4), (df_clust2, df_clust3, df_clust4)

names(df_clust2)
#View(summary(hgOver, pvalue=0.01))
```

##Clusterization with large cluster excluded
```{r E. coli sidd, warning=FALSE, message=FALSE, cache=TRUE}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]
matplot(sidds_ecoli,ty='l',lty =1, col=1, lwd =0.5)

sidds_ecoli_1500 <- sidds_ecoli[750:2250,-1]#row 1 nreeds to be removed since it is a rownumbers

ward_ecoli_1500sidd<- hclust.vector(t(sidds_ecoli_1500), method = 'ward') 
#plot(ward_ecoli_1500sidd, labels = F)
cut_ecoli_1500sidd <- cutree(ward_ecoli_1500sidd, k = 4)
#        table(cut_ecoli_1500sidd)
#removing  prmoters of cluster 2 (largest)

clusters_2_3_4_numbers <- which(cut_ecoli_1500sidd!=1)
#str(clusters_2_3_4_numbers)

# 3 small clusters plotted together

matplot(sidds_ecoli_1500[,clusters_2_3_4_numbers],ty='l',lty =1, col=1, lwd =0.5)

#clusterization of the 3 clusters profiles only

sidds_ecoli_small_clusters <- sidds_ecoli_1500[,clusters_2_3_4_numbers]
ward_ecoli_small_clusters <- hclust.vector(t(sidds_ecoli_1500[,clusters_2_3_4_numbers]), method = 'ward')

#plot(ward_ecoli_small_clusters, labels = F)


#cur and plot the new clusterization results
cuplcl(data = t(sidds_ecoli_small_clusters), method = 'ward', k = 3, main = 'E. coli SIDD profiles: 3 small clusters')
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```


```{r Synechocystis 1500 nts long sidd1, warning=FALSE, message=FALSE, cache=TRUE}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_syn,ty='l',lty =1, col=1, lwd =0.5)

ward_synechocystis_1500sidd<- hclust.vector(t(sidds_syn), method = 'ward')
#plot(ward_synechocystis_1500sidd, labels = F)
cut_synechocystis_1500sidd <- cutree(ward_synechocystis_1500sidd, k = 4)


#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}


sidds_synechocystis_1500 <- sidds_syn[750:2250,-1]#row 1 nreeds to be removed since it is a rownumbers

ward_sidds_synechocystis_1500 <- hclust.vector(t(sidds_synechocystis_1500), method = 'ward') 
#plot(ward_sidds_synechocystis_1500, labels = F)
cuplcl(data = t(sidds_synechocystis_1500), method = 'ward', k = 4, main = 'Synechocystis sp., SIDD')


cut_sidds_synechocystis_1500 <- cutree(ward_sidds_synechocystis_1500, k = 4)
#table(cut_sidds_synechocystis_1500)
#removing  prmoters of cluster 2 (largest)

synechocystis_clusters_2_3_4_numbers <- which(cut_sidds_synechocystis_1500!=4)
#str(synechocystis_clusters_2_3_4_numbers)

# 3 small clusters plotted together


`Sequence (nts)` <- -750:750
matplot(`Sequence (nts)`, sidds_synechocystis_1500[,synechocystis_clusters_2_3_4_numbers],ty='l',lty =1, col=1, lwd =0.5, main = 'Synechocystis sp., 3 small clusters SIDD\n (66 / 634) promoters', ylab =  'Opening probability')

#clusterization of the 3 clusters profiles only

sidds_synechocystis_small_clusters <- sidds_synechocystis_1500[,synechocystis_clusters_2_3_4_numbers]
ward_sidds_synechocystis_small_clusters <- hclust.vector(t(sidds_synechocystis_1500[,synechocystis_clusters_2_3_4_numbers]), method = 'ward')

#plot(ward_sidds_synechocystis_small_clusters, labels = F)


#cur and plot the new clusterization results
cuplcl(data = t(sidds_synechocystis_1500[,synechocystis_clusters_2_3_4_numbers]), method = 'ward', k = 3, main = 'Synechocystis sp., 3 small clusters SIDD')

#sums by rows for reduced (3 small clusters only) dataset


`Sequence (nts)` <- -750:750

sidds_sums_on_positions_synechocystis_3_clusters <- rowSums(sidds_synechocystis_1500[,synechocystis_clusters_2_3_4_numbers])
matplot(`Sequence (nts)`, sidds_sums_on_positions_synechocystis_3_clusters ,ty='l',lty =1, col=1, lwd = 1, main = 'Synechocystis sp., 3 small clusters SIDD\n (66 / 634) promoters', ylab =  'Opening probability')

library(diptest)
diptest::dip.test(x = sidds_sums_on_positions_synechocystis_3_clusters, simulate.p.value = T, B = 10000)
#and for complete set of Synechocystis profiles

sidds_sums_on_positions_synechocystis_all_clusters <- rowSums(sidds_synechocystis_1500)
matplot(`Sequence (nts)`, sidds_sums_on_positions_synechocystis_all_clusters ,ty='l',lty =1, col=1, lwd = 1, main = 'Synechocystis sp., 3 small clusters SIDD\n (66 / 634) promoters', ylab =  'Opening probability')

```
