---
title: "Synechocystis_vs_E.coli"
author: "Mikhail Orlov"
date: '6 ноября 2017 г '
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Clearing workspace, setting working directory and loading R packages
```{r}
rm(list=ls())

#setwd('/home/mikhail/Documents/Script_2016_all/PCA/')

library(R.matlab)
library(seqinr)
library(factoextra)
#library(data.table)
library(caret)
#library(doMC)
library(Biostrings)
library(ape)
library(reldna)
library(dendextend)
```
Reading in E. coli K12 genome (GenBank Accesion U00096.2), creating reverse complement genome and tranformation genome into character form 
```{r Reading in E. coli genome}
e.coli_U00096.2<-paste0(read.GenBank(access.nb = 'U00096.2', as.character = T)$U00096.2, collapse = '')
reverseComplement_e.coli_U00096.2<-as.character(reverseComplement(DNAString(e.coli_U00096.2)))

e.coli_U00096.2_char<-unlist(strsplit(e.coli_U00096.2, ''))
reverseComplement_e.coli_U00096.2_char<-unlist(strsplit(reverseComplement_e.coli_U00096.2, ''))

```

Reading in Synechocystis sp. PCC6803 complete , creating reverse complement genome and tranformation genome into character form 
```{r Reading in Synechocystis sp. PCC6803 complete genome, in supplementary noted genbank acession NC_000911.1}
synechocystis_NC_000911.1 <- toupper(paste0(read.GenBank(access.nb = 'NC_000911.1', as.character = T)$NC_000911.1, collapse = ''))
reverseComplement_synechocystis_NC_000911.1 <- as.character(reverseComplement(DNAString(synechocystis_NC_000911.1)))

synechocystis_NC_000911.1_char<-unlist(strsplit(synechocystis_NC_000911.1, ''))
reverseComplement_synechocystis_NC_000911.1<-unlist(strsplit(reverseComplement_synechocystis_NC_000911.1, ''))

```

Reading in M. gallisepticum strain S6 genome (CP006916.3 assembly) , creating reverse complement genome and tranformation genome into character form 
```{r Reading in M. gallisepticum strain S6 genome (CP006916.3 assembly)}

mycoplasma_CP006916.3<- toupper(paste0(read.GenBank(access.nb = 'CP006916.3', as.character = T)$CP006916.3, collapse = ''))
reverseComplement_mycoplasms_CP006916.3<- as.character(reverseComplement(DNAString(mycoplasma_CP006916.3)))

mycoplasma_CP006916.3_char<-unlist(strsplit(mycoplasma_CP006916.3, ''))
reverseComplement_mycoplasma_CP006916.3<-unlist(strsplit(mycoplasma_CP006916.3, ''))

```



Dynamical properties calculation from within R using octave script translated according to Grinevich et al., 2013
Creating function for dynamical properties calculation

```{r dynchars function }



dynchars<-function(seq, interval_size) {
  if (missing(seq))
    stop("Need to specify sequence (as a vector of chars)")
  
  if (missing(interval_size))
    stop("Need to specify interval size")
  
  if(!is.character(seq))
    stop("Sequence must be a character vector containing A, C, G, T letters only")
  
seq<-toupper(seq)
seq<-c(seq, seq[2:(interval_size)])
  
a<-3.4*10^(-10)
I<-c(7.6, 4.8, 8.2, 4.1)*10^(-44)
K<-c(227, 155, 220, 149)*10^(-20)
V<-c(2.09, 1.43, 3.12, 2.12)*10^(-20)
tau<-c(127, 99, 140, 84)
  
csA<-cumsum(seq=='A') 
csT<-cumsum(seq=='T')
csG<-cumsum(seq=='G')
csC<-cumsum(seq=='C')

countA = csA[interval_size:length(csA)]-c(0, csA[1:(length(csA)-interval_size)])
countT = csT[interval_size:length(csT)]-c(0, csT[1:(length(csT)-interval_size)])
countG = csG[interval_size:length(csG)]-c(0, csG[1:(length(csG)-interval_size)])
countC = csC[interval_size:length(csC)]-c(0, csC[1:(length(csC)-interval_size)])
  
M<-cbind(countA, countT, countG, countC)/interval_size
M_comp<-cbind(countT, countA, countC, countG)/interval_size
M_comp<-apply(t(M_comp),1,rev) 
Is<-as.numeric(M%*%I)#! numeric conversion
Ks<-as.numeric(M%*%K)
Vs<-as.numeric(M%*%V)
  
E01<-(8*(Ks*Vs)^0.5)* 6E23 / 4184
d1<-((Ks*a^2)/Vs)^(0.5)/a;
c1<-(Ks*a^2/Is)^0.5
m1<-E01/c1/6.011E-26
taus1<-as.numeric(M%*%tau) #!as.numeric conversion
gc1 = M[,3] + M[,4]
  
Is<-as.numeric(M%*%I)#! numeric conversion
Ks<-as.numeric(M%*%K)
Vs<-as.numeric(M%*%V)

  
E02<- 8*(Ks*Vs)^0.5  * 6E23 / 4184;
d2<-((Ks*a^2)/Vs)^(0.5)/a;
c2<-(Ks*a^2/Is)^0.5;
m2<-E02/c2/6.011E-26;
taus2<-as.numeric(M_comp%*%tau)
gc2 = M_comp[,3] + M_comp[,4]

dynchars_return<-list(E01=E01, d1=d1, c1=c1, m1=m1, taus1=taus1, gc1=gc1, E02=E02, d2=d2, c2=c2, m2=m2, taus2=taus2, gc2=gc2)
  
return(dynchars_return)
  
}
```

```{r Overall characteristics for three bacteria, cache=TRUE}
whole_e.coli_dynchars <- dynchars(e.coli_U00096.2_char, interval_size = 100)
whole_e.coli_EP <- lseqspline1D(e.coli_U00096.2, bound = c(1, nchar(e.coli_U00096.2)), width = 1, ref =  1)

whole_synechocystis_dynchars <- dynchars(synechocystis_NC_000911.1_char, interval_size = 200)
whole_synechocystis_EP <- lseqspline1D(synechocystis_NC_000911.1, bound = c(1, nchar(synechocystis_NC_000911.1)), width = 1, ref =  1)

whole_mycoplasma_dynchars<- dynchars(mycoplasma_CP006916.3_char, interval_size = 200)
whole_mycoplasma_EP <- lseqspline1D(mycoplasma_CP006916.3, bound = c(1, nchar(mycoplasma_CP006916.3)), width = 1, ref =  1)

par(mfrow = c(3,4))
rbind(sapply(whole_e.coli_dynchars, mean),
sapply(whole_synechocystis_dynchars, mean),
sapply(whole_mycoplasma_dynchars, mean))


par(mfrow = c (1,3))
ylim = c(-0.07, -0.017)

boxplot(whole_e.coli_EP$mpot, ylim = ylim)
boxplot(whole_synechocystis_EP$mpot, ylim = ylim)
boxplot(whole_mycoplasma_EP$mpot, ylim = ylim)


```


```{r}
physdna <- function(seq, interval_size, bound = bound, ref = ref, width = width) {
  #seq must be a character string
  #seq_vec <- unlist(strsplit(seq, split = ''))
                    ##OS dynamics and GC
                    if (missing(seq))
                      stop("Need to specify sequence (as a string)")
                    
                    if (missing(interval_size))
                      stop("Need to specify interval size")
                    
                    if(!is.character(seq))
                      stop("Sequence must be a character vector containing A, C, G, T letters only")
                    
                    #return(dynchars_return)
                    ##bendability
                    library(reldna)
                    bendability_return <- bendability(s = seq, bound = bound, width = width) #width = interval_size
                    electrostatics_return <- lseqspline1D(s = seq, bound = bound, width = width, ref = ref)$mpot
                    risem_return <- dnaGeom(seq)$risem
                    
                    
                    #OS dynamics
                    #seq further used as a character vector
                    seq <- strsplit(seq, split = '')[[1]]
                    seq<-toupper(seq)
                    seq<-c(seq, seq[2:(interval_size)])
                    
                    a<-3.4*10^(-10)
                    I<-c(7.6, 4.8, 8.2, 4.1)*10^(-44)
                    K<-c(227, 155, 220, 149)*10^(-20)
                    V<-c(2.09, 1.43, 3.12, 2.12)*10^(-20)
                    tau<-c(127, 99, 140, 84)
                    
                    csA<-cumsum(seq=='A') 
                    csT<-cumsum(seq=='T')
                    csG<-cumsum(seq=='G')
                    csC<-cumsum(seq=='C')
                    
                    countA = csA[interval_size:length(csA)]-c(0, csA[1:(length(csA)-interval_size)])
                    countT = csT[interval_size:length(csT)]-c(0, csT[1:(length(csT)-interval_size)])
                    countG = csG[interval_size:length(csG)]-c(0, csG[1:(length(csG)-interval_size)])
                    countC = csC[interval_size:length(csC)]-c(0, csC[1:(length(csC)-interval_size)])
                    
                    M<-cbind(countA, countT, countG, countC)/interval_size
                    M_comp<-cbind(countT, countA, countC, countG)/interval_size
                    M_comp<-apply(t(M_comp),1,rev) 
                    Is<-as.numeric(M%*%I)#! numeric conversion
                    Ks<-as.numeric(M%*%K)
                    Vs<-as.numeric(M%*%V)
                    
                    E01<-(8*(Ks*Vs)^0.5)* 6E23 / 4184
                    d1<-((Ks*a^2)/Vs)^(0.5)/a;
                    c1<-(Ks*a^2/Is)^0.5
                    m1<-E01/c1/6.011E-26
                    taus1<-as.numeric(M%*%tau) #!as.numeric conversion
                    gc1 = M[,3] + M[,4]
                    
                    Is<-as.numeric(M%*%I)#! numeric conversion
                    Ks<-as.numeric(M%*%K)
                    Vs<-as.numeric(M%*%V)
                    
                    
                    E02<- 8*(Ks*Vs)^0.5  * 6E23 / 4184;
                    d2<-((Ks*a^2)/Vs)^(0.5)/a;
                    c2<-(Ks*a^2/Is)^0.5;
                    m2<-E02/c2/6.011E-26;
                    taus2<-as.numeric(M_comp%*%tau)
                    gc2 = M_comp[,3] + M_comp[,4]
                    
                    #dynchars_return<-list(E01=E01, d1=d1, c1=c1, m1=m1, taus1=taus1, gc1=gc1, E02=E02, d2=d2, c2=c2, m2=m2, taus2=taus2, gc2=gc2)
                    
                    
                    return(physdna_return <- list(E01=E01, d1=d1, c1=c1, m1=m1, taus1=taus1, gc1=gc1, E02=E02, d2=d2, c2=c2, m2=m2, taus2=taus2, gc2=gc2, bendability_return = bendability_return, electrostatics_return = electrostatics_return, risem_return = risem_return))
}


```

Calculation the properties for a given genome using sliding window 200 nt

```{r Whole genome}

```


Loading data on sequences of different types (promoters, non-promoters, genes, islands, and lowscore) from .Rdata files (must be copied separetely)
```{r}

# #load('/home/mikhail/Documents/Script_2016_all/spline_dataset_pro.Rdata')
load('/home/jane/Документы/Misha/Lab/mol_a-16/spline_dataset_pro.Rdata')
#load('spline_dataset_notpro.Rdata')
#load('spline_dataset_gen.Rdata')
#load('spline_dataset_isl.Rdata')
#load('dataset_lowscore.Rdata')

##dataset_notpro[[1488]]<-NULL # to close to the left flank
```


Check-up for reverse strand
```{r, echo=FALSE}
gregexpr(toupper(dataset_pro$aaeRp10$seq), reverseComplement_e.coli_U00096.2)
nchar(e.coli_U00096.2)-dataset_pro$aaeRp10$tss-270
```
Extracting data on all promoters and on experimentaly found ones - including previosely calculated electrostatic potential profiles
```{r}
pro_names<-names(dataset_pro)
tsss<-c()
seqs<-c()
exp_strands<-c()
exp_tsss<-c()
exp_EP_check<-c() #for experimentally found one
exp_names<-c()

for (i in 1:length(dataset_pro)){
  tsss<-c(tsss, dataset_pro[[i]]$tss)
  seqs<-c(seqs, dataset_pro[[i]]$seq)
  if (dataset_pro[[i]]$evidence=='experimental'){ 
    #exp_proms<-rbind(exp_proms, c(pro_names[i], strands[i], tsss[i], seqs[i]))
    exp_names<-c(exp_names, pro_names[i])
    exp_EP_check<-rbind(exp_EP_check, dataset_pro[[i]]$mpot)  	
    #exp_promoters<-rbind(exp_promoters, c(pro_names[i], strands[i], tsss[i], seqs[i]))
    exp_tsss<-c(exp_tsss, dataset_pro[[i]]$tss) 
    exp_strands<-c(exp_strands, dataset_pro[[i]]$strand)}
}
```

# # # Changing genomic coordinate used for dataset_... - TSS is according to 5'-end for both strands 
```{r}

exp_tsss[which(exp_strands=='reverse')]<-nchar(e.coli_U00096.2)-exp_tsss[which(exp_strands=='reverse')]
#exp_tsss<--exp_tsss
```


```{r strReverse funtcion}
## a useful function: rev() for strings
strReverse <- function(x)
        sapply(lapply(strsplit(x, NULL), rev), paste, collapse="")
```


Creating matrices for data on activation energy ('E01') and size ('d1'). The matrices is 699*201 - 699 promoters sequences, 201 value for a physical property profile
to 5'-end for both strands 


Creating matrices of dynamical properties and GC-content for intervals [-150; 50] nts
```{r E. coli promoters physics, cache=TRUE} 
for (i in c('seqsforward', 'seqsreverse', 'E01forward', 'E01reverse', 'd1forward', 'd1reverse', 'GCforward', 'GCreverse', 'EPforward', 'EPreverse')) {
  assign(i, c())
}
zout<- -480:239


for (i in 1:length(exp_tsss)) {
  if (exp_strands[i]=='forward') {
     seqsforward<-c(seqsforward, substr(e.coli_U00096.2, start = exp_tsss[i] - 150, stop = exp_tsss[i] + 50))
                                    
    E01forward<-rbind(E01forward, whole_e.coli_dynchars$E01[as.numeric(exp_tsss[i]-150):(exp_tsss[i]+50)])
    #aeos2forward<-rbind(aeos2forward, matr1[(as.numericexp_tsss[i]-150):(as.numericexp_tsss[i]+50),2])
    d1forward<-rbind(d1forward, whole_e.coli_dynchars$d1[(as.numeric(exp_tsss[i])-150):(exp_tsss[i]+50)])
    #aeos4forward<-rbind(aeos4forward, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    GCforward<-rbind(GCforward, whole_e.coli_dynchars$gc1[(as.numeric(exp_tsss[i])-150):(as.numeric(exp_tsss[i])+50)])
    
    p<-lseqspline1D(substr(e.coli_U00096.2, exp_tsss[i]-250, exp_tsss[i]+150), bound=c(50, 350), ref=251 )
    EPforward<-rbind(EPforward, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
     } else {
    seqsreverse<-c(seqsreverse, substr(e.coli_U00096.2, start = exp_tsss[i] - 150, stop = exp_tsss[i] + 50))
    E01reverse<-rbind(E01reverse, whole_e.coli_dynchars$E02[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)])
    #aeos2reverse<-rbind(aeos2reverse, matr1[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150),2])
    d1reverse<-rbind(d1reverse, whole_e.coli_dynchars$d2[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)])
    #aeos4reverse<-rbind(aeos4reverse, matr1[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150),4])
    GCreverse<-rbind(GCreverse, (whole_e.coli_dynchars$gc2[(as.numeric(exp_tsss[i])-50):(as.numeric(exp_tsss[i])+150)]))
    
    
    
    p<-lseqspline1D(substr(reverseComplement_e.coli_U00096.2, exp_tsss[i]-250, exp_tsss[i]+150), bound=c(50, 350), ref=251 )
    EPreverse<-rbind(EPreverse, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
    
    }
}


#merging matrices for forward and reverse strands  together
seqs <- toupper(c(seqsforward, sapply(seqsreverse, FUN = strReverse)))
 E01<-rbind( E01forward, t(apply( E01reverse, 1, rev)))
#aeos2<-rbind(aeos2forward, aeos2reverse)
 d1<-rbind( d1forward, t(apply( d1reverse, 1, rev)))
#aeos4<-rbind(aeos4forward, aeos4reverse)
 GC<-rbind( GCforward, t(apply( GCreverse, 1, rev)))

 EP<-rbind( EPforward, t(apply( EPreverse, 1, rev)))
```




```{r Datasets from An experimentally anchored map of transcriptional start sites in the model cyanobacterium Synechocystis sp. PCC6803 Jan Mitschke, 2011}

synechocystis_proms <- (read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/synechocystis_proms_datasets_1015154108_sd01.csv', skip = 14, header = T))

```

```{r Checking whether the strain is correct, eval=F}
tmp_forward <- synechocystis_proms$X1.TSS[3]
tmp_reverse <- synechocystis_proms$X1.TSS[1]
synechocystis_proms$X4.strand[1:10] #3 is forward
synechocystis_proms$X6.observed.sequence[1]

substr(reverseComplement_synechocystis_NC_000911.1, start = tmp_reverse - 15, stop = tmp_reverse+15)

#synechocystis_proms$X6.observed.sequence[3]
#substr(synechocystis_NC_000911.1, start = tmp_forward - 15, stop = tmp_forward+15)


which.synechocystis_proms.fwd <- which(synechocystis_proms$X4.strand=='fwd')
which.synechocystis_proms.rev <- which(synechocystis_proms$X4.strand=='rev')

greped_forward <- sapply(synechocystis_proms$X6.observed.sequence[which.synechocystis_proms.fwd], FUN = function(x) {grep(x, synechocystis_NC_000911.1)})
length(greped_forward)
greped_reverse <- sapply(synechocystis_proms$X6.observed.sequence[which.synechocystis_proms.rev], FUN = function(x) {grep(x, synechocystis_NC_000911.1)})
length(greped_reverse)
length(synechocystis_proms$X1.TSS)
```

Extracting data on all promoters and on experimentaly found ones - including previosely calculated electrostatic potential profiles for Synechosystis
Creating matrices for data on activation energy ('E01') and size ('d1'). The matrices is 699*201 - 699 promoters sequences, 201 value for a physical property profile
to 5'-end for both strands 


Creating matrices of dynamical properties and GC-content for intervals [-150; 50] nts
```{r Synechocystis proms physics, cache=TRUE} 


for (i in c('synechocystis_seqsforward', 'synechocystis_seqsreverse', 'synechocystis_E01forward', 'synechocystis_E01reverse', 'synechocystis_d1forward', 'synechocystis_d1reverse', 'synechocystis_GCforward', 'synechocystis_GCreverse', 'synechocystis_EPforward', 'synechocystis_EPreverse')) {
  assign(i, c())
}
zout<- -480:239
interval_size <- 50
for (i in 1:length(synechocystis_proms$X1.TSS)) {
  if (synechocystis_proms$X4.strand[i]=='fwd') {
    
    synechocystis_seqsforward<-c(synechocystis_seqsforward, substr(synechocystis_NC_000911.1, start = synechocystis_proms$X1.TSS[i] - 150, stop = synechocystis_proms$X1.TSS[i] + 50))
                                    
                                    
    synechocystis_E01forward<-rbind(synechocystis_E01forward,
dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$E01)                       
    synechocystis_d1forward<-rbind(synechocystis_d1forward, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$d1)                       
    #aeos4forward<-rbind(aeos4forward, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    synechocystis_GCforward<-rbind(synechocystis_GCforward, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-150):(synechocystis_proms$X1.TSS[i]+50)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(synechocystis_NC_000911.1, synechocystis_proms$X1.TSS[i]-250, synechocystis_proms$X1.TSS[i]+150), bound=c(50, 350), ref=251 )
    synechocystis_EPforward<-rbind(synechocystis_EPforward, 
                        ##mpot_whole_genome_forward$mpot[mpot_whole_genome_forward$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
     } else  {
    synechocystis_seqsreverse<-c(synechocystis_seqsreverse, substr(synechocystis_NC_000911.1, start = synechocystis_proms$X1.TSS[i] - 50, stop = synechocystis_proms$X1.TSS[i] + 150))
       
    synechocystis_E01reverse<-rbind(synechocystis_E01reverse,
dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$E01)                       
    synechocystis_d1reverse<-rbind(synechocystis_d1reverse, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$d1)                       
    #aeos4reverse<-rbind(aeos4reverse, matr1[(as.numeric(exp_proms[i,3])-150):(as.numeric(exp_proms[i,3])+50),4])            
    synechocystis_GCreverse<-rbind(synechocystis_GCreverse, dynchars(synechocystis_NC_000911.1_char[(synechocystis_proms$X1.TSS[i]-50):(synechocystis_proms$X1.TSS[i]+150)], interval_size = interval_size)$gc1)                       
    
    p<-lseqspline1D(substr(synechocystis_NC_000911.1, synechocystis_proms$X1.TSS[i]-150, synechocystis_proms$X1.TSS[i]+250), bound=c(50, 350), ref=251 )
    synechocystis_EPreverse<-rbind(synechocystis_EPreverse, 
                        ##mpot_whole_genome_reverse$mpot[mpot_whole_genome_reverse$x %in% (exp_tsss[i]-150):(exp_tsss[i]+50)])
                        p$mpot[p$x %in% zout])
    
    
    
    
    }
}


#merging matrices for forward and reverse strands  together
synechocystis_seqs <- toupper(c(synechocystis_seqsforward, sapply(synechocystis_seqsreverse, FUN = strReverse)))
synechocystis_E01<-rbind(synechocystis_E01forward, t(apply(synechocystis_E01reverse, 1, rev)))
#aeos2<-rbind(aeos2forward, aeos2reverse)
synechocystis_d1<-rbind(synechocystis_d1forward, t(apply(synechocystis_d1reverse, 1, rev)))
#aeos4<-rbind(aeos4forward, aeos4reverse)
synechocystis_GC<-rbind(synechocystis_GCforward, t(apply(synechocystis_GCreverse, 1, rev)))

synechocystis_EP<-rbind(synechocystis_EPforward, t(apply(synechocystis_EPreverse, 1, rev)))

```



















```{r  Reading in M. gallisepticum strain S6 promoters }

mycoplasma_proms <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/mycoplasma_gallisepticum_fisunov_2016.csv', header = T)

plot(seq_along(mycoplasma_CP006916.3_char), type = 'n')
abline(v = mycoplasma_proms$Position, col = 'red')

table(mycoplasma_proms$Strand)
```

```{r Checking whether the strain is correct 1, eval=F}
tmp_forward <- synechocystis_proms$X1.TSS[3]
tmp_reverse <- synechocystis_proms$X1.TSS[1]
synechocystis_proms$X4.strand[1:10] #3 is forward
synechocystis_proms$X6.observed.sequence[1]

substr(reverseComplement_synechocystis_NC_000911.1, start = tmp_reverse - 15, stop = tmp_reverse+15)

#synechocystis_proms$X6.observed.sequence[3]
#substr(synechocystis_NC_000911.1, start = tmp_forward - 15, stop = tmp_forward+15)


which.synechocystis_proms.fwd <- which(synechocystis_proms$X4.strand=='fwd')
which.synechocystis_proms.rev <- which(synechocystis_proms$X4.strand=='rev')

greped_forward <- sapply(synechocystis_proms$X6.observed.sequence[which.synechocystis_proms.fwd], FUN = function(x) {grep(x, synechocystis_NC_000911.1)})
length(greped_forward)
greped_reverse <- sapply(synechocystis_proms$X6.observed.sequence[which.synechocystis_proms.rev], FUN = function(x) {grep(x, synechocystis_NC_000911.1)})
length(greped_reverse)
length(synechocystis_proms$X1.TSS)
```

Extracting data on all promoters and on experimentaly found ones - including previosely calculated electrostatic potential profiles for Synechosystis
Creating matrices for data on activation energy ('E01') and size ('d1'). The matrices is 699*201 - 699 promoters sequences, 201 value for a physical property profile
to 5'-end for both strands 




```{r Plotting characteristics profiles for Cynechocystis}

par(mfrow = c(2,2))

matplot(synechocystis_E01, type = 'l', col = 1)
matplot(synechocystis_d1, type = 'l', col = 1)
matplot(synechocystis_GC, type = 'l', col = 1)
matplot(synechocystis_EP, type = 'l', col = 1)

```


```{r string distance}
library(Biostrings)



synechocystis_e.coli_proms_strings <- sapply(c(synechocystis_seqs, seqs),  function(x) {paste0(x, collapse = '')})
synechocystis_e.coli_proms_DNAStringSet <- DNAStringSet(synechocystis_e.coli_proms_strings)

synechocystis_e.coli_proms_stringdist <- stringDist(synechocystis_e.coli_proms_DNAStringSet, method = 'levenshtein')

ward_synechocystis_e.coli_proms_stringdist <- hclust(synechocystis_e.coli_proms_stringdist, method = 'ward.D2')

plot(ward_synechocystis_e.coli_proms_stringdist, labels = F)

colvec_synechocystis_e.coli_proms <- c(rep('darkgreen', length(synechocystis_proms$X1.TSS)),
rep('orange', length(exp_tsss)))
ward_synechocystis_e.coli_proms_stringdist %>% as.dendrogram %>% color_branches(k = length(ward_synechocystis_e.coli_proms_stringdist$order), col = colvec_synechocystis_e.coli_proms[order.dendrogram(ward_synechocystis_e.coli_proms_stringdist %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_synechocystis_e.coli_proms_stringdist$height)*0.2)%>% circlize_dendrogram(labels=F, dend_track_height = 0.95) -> to_plot_ward_synechocystis_e.coli_proms_stringdist


#plot(to_plot_ward_synechocystis_e.coli_proms_stringdist, main = 'E. coli and Synechocystis promoters: DNA Levenshtein distance')
```



```{r two bacteria promoters mixture - all the characteristics}
library(fastcluster)
library(dendextend)
cex.main = 2


par(mfrow = c(5,1),
    mar = rep(0.8,4))

to_plot_wards <- list()
for (i in c('E01', 'd1', 'GC', 'EP')){
  both_tmp <- rbind(get(i), get(paste0('synechocystis_', i)))
  ward_both_tmp<- hclust.vector(scale(both_tmp, center = T, scale = T), method = 'ward')

  colvec_ward_both_tmp <- c(rep('orange', length(exp_tsss)),rep('darkgreen',length(synechocystis_proms$X1.TSS)))[order.dendrogram(ward_both_tmp %>% as.dendrogram )]

  ward_both_tmp %>% as.dendrogram %>% color_branches(k = length(ward_both_tmp$order), col = colvec_ward_both_tmp) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_both_tmp$height)*.2) -> to_plot_ward_both_tmp

  plot(to_plot_ward_both_tmp, labels = F, main = paste0('E.coli and Synechocystis promoters (', i, ')'), cex.main = cex.main )
  
  assign(paste0('to_plot_wards', i), ward_both_tmp)
}

plot(to_plot_ward_synechocystis_e.coli_proms_stringdist, main = 'E. coli and Synechocystis promoters: DNA Levenshtein distance', cex.main = cex.main )
```



```{r substring occurences}

library(Biostrings)


#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(synechocystis_seqs, seqs, synechocystis_NC_000911.1, e.coli_U00096.2)){
  print(length(i))
  tmp <- DNAString(paste0(unlist(i), collapse = ''))
  
  #tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}



colnames(all_seqs_oligos_freq) <- c('Synechocystis promoters', 'E. coli promoters', 'Synechocystis sp., complete genome', 'E. coli, complete genome' )
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]


par(mfrow = c(1,2))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
  barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

```


```{r hexamers occurences PCA, pca 1-2}
PCA <- prcomp(all_seqs_oligos_freq[,1:2], scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')

label.size <- 1.7
#library(ggfortify)
#autoplot(PCA,
    #     shape = FALSE,
   #      loadings = TRUE,
  #       label = TRUE,
 #        loadings.label = TRUE,
#         label.size = label.size)+
 # theme_bw()

library(factoextra)

fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=0.6)
  
```

```{r hexamers occurences PCA, pca 1-3}
PCA <- prcomp(all_seqs_oligos_freq[,c(1,3)], scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')


fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=0.6)
  
```

```{r hexamers occurences PCA, pca 4-3}
PCA <- prcomp(all_seqs_oligos_freq[,c(3,4)], scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')


fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=0.6)
  
```


```{r Other text representation}

#panel.cor function
panel.cor <- function(x, y, digits=2, prefix="", cex.cor) 
{
    usr <- par("usr"); on.exit(par(usr)) 
    par(usr = c(0, 1, 0, 1)) 
    r <- abs(cor(x, y)) 
    txt <- format(c(r, 0.123456789), digits=digits)[1] 
    txt <- paste(prefix, txt, sep="") 
    if(missing(cex.cor)) cex <- 0.8/strwidth(txt) 
 
    test <- cor.test(x,y) 
    # borrowed from printCoefmat
    Signif <- symnum(test$p.value, corr = FALSE, na = FALSE, 
                  cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                  symbols = c("***", "**", "*", ".", " ")) 
 
    text(0.5, 0.5, txt, cex = cex * r) 
    text(.8, .8, Signif, cex=cex, col=2) 
}

pairs(all_seqs_oligos_freq, lower.panel=panel.smooth, upper.panel=panel.cor)


par(mfrow = c(1,4))
apply(all_seqs_oligos_freq, 2, function(x) {
  boxplot(x, ylim = range(all_seqs_oligos_freq))
  }
  )

for (i in 1:ncol(all_seqs_oligos_freq)){
  assign(paste0('lm', i), lm(all_seqs_oligos_freq[,i] ~ seq_along(all_seqs_oligos_freq[,i])))
}

par(mfrow = c(4,5),
    mar = rep(2,4))
plotnames <- c('Synechocystis promoters', 'E. coli promoters', 'Synechocystis sp. genome', 'E. coli genome')
for (i in 1:ncol(all_seqs_oligos_freq)){
 
  plot(1, type="n", axes=F, xlab="", ylab="")
  mtext(plotnames[i], side = 3, col = 2, cex= 0.7, line = -3)

  plot(get(paste0('lm', i)))
 
}
#boxplot(all_seqs_oligos_freq[,1])
#identify(rep(1, length(all_seqs_oligos_freq[,1])), all_seqs_oligos_freq[,1], labels = #seq_along(all_seqs_oligos_freq[,1]))
```

```{r 6 plots}

par(mfrow = c(3,2),
    mar = rep(1,4))

to_plot_wards <- list()
for (i in c('E01', 'd1', 'GC', 'EP')){
  both_tmp <- rbind(get(i), get(paste0('synechocystis_', i)))
  ward_both_tmp<- hclust.vector(scale(both_tmp, center = T, scale = T), method = 'ward')

  colvec_ward_both_tmp <- c(rep('orange', length(exp_tsss)),rep('darkgreen',length(synechocystis_proms$X1.TSS)))[order.dendrogram(ward_both_tmp %>% as.dendrogram )]

  ward_both_tmp %>% as.dendrogram %>% color_branches(k = length(ward_both_tmp$order), col = colvec_ward_both_tmp) %>% set('labels_cex', 0.000001) %>% raise.dendrogram(max(ward_both_tmp$height)*.2) -> to_plot_ward_both_tmp

  plot(to_plot_ward_both_tmp, labels = F, main = paste0('E.coli and Synechocystis promoters (', i, ')'), cex.main = cex.main )
  
  assign(paste0('to_plot_wards', i), ward_both_tmp)
}

plot(to_plot_ward_synechocystis_e.coli_proms_stringdist, main = 'E. coli and Synechocystis promoters: DNA Levenshtein distance', cex.main = cex.main )

##load('/home/mikhail/Documents/mol_a_2017/sidds_ecoli.rda')
##load('/home/mikhail/Documents/mol_a_2017/sidds_synechocystis.rda')

sidds_ecoli <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_ecoli.csv')
sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis.csv')


library(fastcluster)
synechocystis_e.coli_proms_sidds <- cbind(sidds_ecoli, sidds_syn)


ward_synechocystis_e.coli_proms_sidd <- hclust.vector(t(synechocystis_e.coli_proms_sidds), method = 'ward')

#plot(ward_synechocystis_e.coli_proms_sidd, labels = F)

colvec_ward_synechocystis_e.coli_proms_sidd <- c(rep('darkgreen', length(synechocystis_proms$X1.TSS)),
rep('orange', length(exp_tsss)))
ward_synechocystis_e.coli_proms_sidd %>% as.dendrogram %>% color_branches(k = length(ward_synechocystis_e.coli_proms_sidd$order), col = colvec_ward_synechocystis_e.coli_proms_sidd[order.dendrogram(ward_synechocystis_e.coli_proms_sidd %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(ward_synechocystis_e.coli_proms_sidd$height)*0.2) -> to_plot_ward_synechocystis_e.coli_proms_sidd #circlize_dendrogram(labels=F, dend_track_height = 0.95) 
  

plot(to_plot_ward_synechocystis_e.coli_proms_sidd, labels = F)

```


```{r cuplcl usage}
#source('/home/mikhail/Documents/Script_2016_all/cut_and_plot_clusters.R')

#function to cut heirarchical clusterization results at given clusters numer and plot all pfrofiles from each

cuplcl <- function(data, method, k) {
library(fastcluster)
library(dendextend)
hclust <- hclust.vector(data, method = method)
cut_hclust<-cutree(hclust, k=k)
m <- rbind(rep(1, k), c(2:(k+1)))
layout(m)
par(mar = c(1, 1, 0, 0))

dend <- color_branches(as.dendrogram(hclust), col = cut_hclust[order.dendrogram(as.dendrogram(hclust))])
plot(dend%>%set('labels_cex', NA))

for (i in unique(cut_hclust[order.dendrogram(as.dendrogram(hclust))])) {
matplot(t(data[which(cut_hclust==i),]), type='l', col=i)

print(c(i, names(which(cut_hclust==i))))
}
}
```




```{r #tanglegrams3, cache=T}

#labels(to_plot_wards[[1]]) <- labels(to_plot_wards[[2]]) <- as.character(labels(to_plot_wards[[2]]))


#for (i in c('E01', 'd1', 'GC', 'EP')){
#for (j in c('E01', 'd1', 'GC', 'EP')){

 #   if (i == j) {next} else {
      
#tanglegram(to_plot_wardsE01, to_plot_wardsd1, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsE01, to_plot_wardsGC, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsE01, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)


#tanglegram(to_plot_wardsd1, to_plot_wardsGC, lwd = 0.2, lab.cex = 1E-10)
#tanglegram(to_plot_wardsd1, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)

#tanglegram(to_plot_wardsGC, to_plot_wardsEP, lwd = 0.2, lab.cex = 1E-10)
  #plot(to_plot_ward_both_tmp, labels = F, main = paste0('E.coli and Synechocystis promoters (', i, ')'), cex.main = cex.main )
  #  }
  #}
#}
```


```{r E. coli 1500 nts long sidd}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_ecoli,ty='l',lty =1, col=1, lwd =0.5)

ward_ecoli_1500sidd<- hclust.vector(t(sidds_ecoli), method = 'ward')
plot(ward_ecoli_1500sidd, labels = F)
cut_ecoli_1500sidd <- cutree(ward_ecoli_1500sidd, k = 4)

cuplcl(data = sidds_ecoli, method = 'ward', k = 4)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```


```{r Synechocystis 1500 nts long sidd}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_syn,ty='l',lty =1, col=1, lwd =0.5)

ward_synechocystis_1500sidd<- hclust.vector(t(sidds_syn), method = 'ward')
plot(ward_synechocystis_1500sidd, labels = F)
cut_synechocystis_1500sidd <- cutree(ward_synechocystis_1500sidd, k = 4)

cuplcl(data = sidds_syn, method = 'ward', k = 4)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```



```{r Whole-genome SIDD vectors}
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_2500_500.rda')
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_5000_1000.rda')
load('/home/jane/Документы/Misha/Lab/mol_a-16/spline_dataset_pro.Rdata')

exp_pro_numbers<- which(sapply(dataset_pro, function(x) {x$evidence =='experimental' })==T)
exp_pro_tsss <- sapply(dataset_pro, function(x){x$tss})[exp_pro_numbers]


par(mfrow = c(2,1))
interv <- 1:100000
plot(whole_e.coli_sidd_2500_500[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )
plot(whole_e.coli_sidd_5000_1000[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )

exp_pro_sidd_2500_500 <- sapply(exp_pro_tsss, function(x) {whole_e.coli_sidd_2500_500[(x-1000):(x+1000)]})
matplot(exp_pro_sidd_2500_500, type = 'l', col =1, lwd =0.3)

exp_pro_sidd_5000_1000 <- sapply(exp_pro_tsss, function(x) {whole_e.coli_sidd_5000_1000[(x-1000):(x+1000)]})
matplot(exp_pro_sidd_5000_1000, type = 'l', col =1, lwd =0.3)
```


```{r Rollaply max}
library(zoo)

rollmax_whole_e.coli_sidd_2500_500 <- rollmax(x = whole_e.coli_sidd_2500_500, k = 250)

rollmax_whole_e.coli_sidd_5000_1000 <- rollmax(x = whole_e.coli_sidd_5000_1000, k = 250)


par(mfrow = c(2,1),
    mar = rep(0.7,4))
interv <- 1:50000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )
plot(rollmax_whole_e.coli_sidd_5000_1000[interv], type = 'l')
abline(v = exp_pro_tsss, col = 2 )

which_rollmax_whole_e.coli_sidd_2500_500_0.95 <- which(rollmax_whole_e.coli_sidd_2500_500>0.95)
which_rollmax_whole_e.coli_sidd_5000_1000_0.95 <- which(rollmax_whole_e.coli_sidd_5000_1000>0.95)                                              
#how big is a moltem part of the genome (nts)
length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)
length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)

par(mfrow = c(2,1),
    mar = rep(0.7,4))
interv <- 1:500000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_2500_500_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)), pch = '|', col = 2)
abline(v = exp_pro_tsss, col = 'darkgreen' )

plot(rollmax_whole_e.coli_sidd_5000_1000[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_5000_1000_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)), pch = '|', col = 2)
abline(v = exp_pro_tsss, col = 'darkgreen' )

#which regions are melted for different split sets
interv <- 300000:400000
plot(rollmax_whole_e.coli_sidd_2500_500[interv], type = 'n')
points(x = which_rollmax_whole_e.coli_sidd_2500_500_0.95, y = rep(0.95, length(which_rollmax_whole_e.coli_sidd_2500_500_0.95)), pch = '|', col = 2)
text(x = 0, y =0.95, labels = '2500-500\n fragments', cex = 0.7)
points(x = which_rollmax_whole_e.coli_sidd_5000_1000_0.95, y = rep(0.75, length(which_rollmax_whole_e.coli_sidd_5000_1000_0.95)), pch = '|', col = 2)
text(x = 0, y =0.75, labels = '5000-1000\n fragments', cex = 0.7)
points(x = exp_pro_tsss, y = rep(0.55, length(exp_pro_tsss)), pch = '|', col = 'darkblue')
text(x = 0, y =0.55, labels = 'Experimentally\n evidenced\n promoters', cex = 0.7)

```
Column 1 contains numbers and is removed 
```{r CHecking bimodality for SIDD for all promoters}

interv <- 750:2250
interv <- 500:2500
####interv <- 900:2100
####interv <- 1000:2000

sums_sidds_ecoli <- apply(sidds_ecoli[,-1], 2, function(x) {sum(x[interv])})

sums_sidds_syn <- apply(sidds_syn[,-1], 2, function(x) {sum(x[interv])})


#svg('/home/jane/Документы/Misha/mol_a_2018/different_bacteria_physics_an_text/densities_auc_sidd_2_bacteria.svg', height = 10, width = 10)
par(mfrow = c(2,1))
hist(sums_sidds_ecoli, breaks = 15, freq = F, xlab = 'AUC for 1500 nts-long SIDD profile', main = 'Histogram of E. coli Promoters AUC SIDD Data\n with Kernel Density Plot')
lines(density(sums_sidds_ecoli, na.rm = T))


hist(sums_sidds_syn, breaks = 15, freq = F, xlab = 'AUC for 1500 nts-long SIDD profile', main = 'Histogram of Synechocystis sp. Promoters AUC SIDD Data\n with Kernel Density Plot')
lines(density(sums_sidds_syn, na.rm = T))

#dev.off()

library(diptest)
diptest_sums_sidds_ecoli <- dip.test(sums_sidds_ecoli, simulate.p.value = T, B = 20000) #D = 0.055838, p-value < 2.2e-16 for 1500 nts -long interval
# D = 0.029814, p-value = 1e-04 for 2000 nts
diptest_sums_sidds_ecoli

diptest_sums_sidds_syn<- dip.test(sums_sidds_syn, simulate.p.value = T, B = 20000) #D = 0.037924, p-value < 2.2e-16 for 1500 nts -long interval
#D = 0.045919, p-value < 2.2e-16 for 2000 nts
diptest_sums_sidds_syn
```


```{r if dynamic properties ditributed unimodally, eval=FALSE}


funcs <- c('sum', 'mean', 'median', 'max')

for (i in funcs){
  assign(paste0(i,' _E01_e.coli_promoters'), apply(E01, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_E01_e.coli_promoters'))))
}


for (i in funcs){
  assign(paste0(i,'_d1_e.coli_promoters'), apply(d1, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_d1_e.coli_promoters'))))
}


for (i in funcs){
  assign(paste0(i,'_GC_e.coli_promoters'), apply(GC, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_GC_e.coli_promoters'))))
}

for (i in funcs){
  assign(paste0(i,'_EP_e.coli_promoters'), apply(GC, MARGIN = 1, FUN = i))
}


par(mfrow = c(2,2))
for (i in funcs){
  plot(density(get(paste0(i,'_GC_e.coli_promoters'))))
}


```

```{r GO statistics using GOstats}
library(GOstats)
library(Category)
library(DBI)
library(org.EcK12.eg.db)

load('/home/jane/Документы/Misha/mol_a_2018/different_bacteria_physics_an_text/ecoli_sidd_clustering_df_to_go.rda')

entrezUniverse <- as.character(dfGO$Entrezid)
selectedEntrezIds_clust2 <- as.character(dfGO$Entrezid[which(dfGO$group == 2)])
selectedEntrezIds_clust3 <- as.character(dfGO$Entrezid[which(dfGO$group == 3)])
selectedEntrezIds_clust4 <- as.character(dfGO$Entrezid[which(dfGO$group == 4)])

#setting cutoff value
hgCutoff <- 0.001
#hgCutoff <- 0.1

#selecting onthology

ontology <- 'BP'
#ontology <- 'MF'

params_clust2 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust2,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

params_clust3 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust3,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

params_clust4 <- new("GOHyperGParams",
 geneIds=selectedEntrezIds_clust4,
 universeGeneIds=entrezUniverse,
 annotation="org.EcK12.eg.db",
 ontology=ontology,
 pvalueCutoff=hgCutoff,
 conditional=FALSE,
 testDirection="over")

#We would also like to perform a conditional test.  Instead of having to de ne a new
#GO-HyperGParamsinstance  by  hand,  we  can  create  a  copy  and  update  just  the  parameter  of interest.
#paramsCond <- params
#conditional(paramsCond) <- TRUE
#hgCondOver <- hyperGTest(paramsCond)



hgOver_clust2 <- hyperGTest(params_clust2)
hgOver_clust3 <- hyperGTest(params_clust3)
hgOver_clust4 <- hyperGTest(params_clust4)



df_clust2 <- summary(hgOver_clust2)
df_clust3 <- summary(hgOver_clust3)
df_clust4 <- summary(hgOver_clust4)

########cbind(rbind(df_clust2, df_clust3, df_clust4), (df_clust2, df_clust3, df_clust4)

names(df_clust2)
#View(summary(hgOver, pvalue=0.01))
```

##Clusterization with large cluster excluded
```{r E. coli sidd}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]
matplot(sidds_ecoli,ty='l',lty =1, col=1, lwd =0.5)

sidds_ecoli_1500 <- sidds_ecoli[750:2250,-1]#row 1 nreeds to be removed since it is a rownumbers

ward_ecoli_1500sidd<- hclust.vector(t(sidds_ecoli_1500), method = 'ward') 
plot(ward_ecoli_1500sidd, labels = F)
cut_ecoli_1500sidd <- cutree(ward_ecoli_1500sidd, k = 4)
#        table(cut_ecoli_1500sidd)
#removing  prmoters of cluster 2 (largest)

clusters_2_3_4_numbers <- which(cut_ecoli_1500sidd!=2)
str(clusters_2_3_4_numbers)

# 3 small clusters plotted together

matplot(sidds_ecoli_1500[,clusters_2_3_4_numbers],ty='l',lty =1, col=1, lwd =0.5)

#clusterization of the 3 clusters profiles only

sidds_ecoli_small_clusters <- sidds_ecoli_1500[,clusters_2_3_4_numbers]
ward_ecoli_small_clusters <- hclust.vector(t(sidds_ecoli_1500[,clusters_2_3_4_numbers]), method = 'ward')

plot(ward_ecoli_small_clusters, labels = F)


#cur and plot the new clusterization results
cuplcl(data = t(sidds_ecoli_small_clusters), method = 'ward', k = 3)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```


```{r Synechocystis 1500 nts long sidd1}
# sidds_syn <- read.csv('https://raw.githubusercontent.com/FVortex/different_bacteria_physics_and_text/master/sidds_synechocystis_1500nts.csv' )[,-1]

matplot(sidds_syn,ty='l',lty =1, col=1, lwd =0.5)

ward_synechocystis_1500sidd<- hclust.vector(t(sidds_syn), method = 'ward')
plot(ward_synechocystis_1500sidd, labels = F)
cut_synechocystis_1500sidd <- cutree(ward_synechocystis_1500sidd, k = 4)

cuplcl(data = sidds_syn, method = 'ward', k = 4)
#`Sequence (nts)` <- -750:750
#par(mfrow = c(3,2))
#plot(ward_synechocystis_1500sidd, labels = F)
#for (i in unique(cut_synechocystis_1500sidd)){
  #matplot(`Sequence (nts)`, sidds_syn[, which(cut_synechocystis_1500sidd==i)], type = 'l', lty = 1, col=i)
 # abline(v = -10:10*100, lty =3, col = 'darkgrey', lwd =2)
#}
```