---
title: "sidd_all_seq_types.Rmd"
author: "Mikhail Orlov"
date: '13 октября 2017 г '
output: pdf_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown SIDD for all sequence types
Reading in E. coli K12 genome (GenBank Accesion U00096.2), creating reverse complement genome and tranformation genome into character form 

```{r}
library(seqinr)
library(Biostrings)
e.coli_U00096.2<-unlist(read.fasta('/home/jane/Документы/Misha/Lab/mol_a-16/Genome_to_dataset_pro_U00096.2.txt', seqonly = T))
reverseComplement_e.coli_U00096.2<-as.character(reverseComplement(DNAString(e.coli_U00096.2)))

load('/home/jane/Загрузки/spline_dataset_pro.Rdata')
load('/home/jane/Загрузки/spline_dataset_notpro.Rdata')
load('/home/jane/Загрузки/spline_dataset_gen.Rdata')
load('/home/jane/Загрузки/spline_dataset_isl.Rdata')
load('/home/jane/Загрузки/dataset_lowscore.Rdata')

load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_5000_1000.rda')


which.experimental <- which(sapply(dataset_pro, function(x) {(x$evidence == 'experimental')}))
exptsss <- sapply(dataset_pro, function(x) {x$tss})[which.experimental]
expstrands <- sapply(dataset_pro, function(x) {x$strand})[which.experimental]

nottsss <- sapply(dataset_notpro, function(x) {x$tss})
notstrands <- sapply(dataset_notpro, function(x) {x$strand})

gentsss <- sapply(dataset_gen, function(x) {x$tss})
genstrands <- sapply(dataset_gen, function(x) {x$strand})

isltsss <- sapply(dataset_isl, function(x) {x$tss})
islstrands <- sapply(dataset_isl, function(x) {x$strand})

lowscoretsss <- sapply(dataset_lowscore, function(x) {x$tss}) #only some are to be used
lowscorestrands <- sapply(dataset_lowscore, function(x) {x$strand})

## a useful function: rev() for strings
strReverse <- function(x)
        sapply(lapply(strsplit(x, NULL), rev), paste, collapse="")

expseqs <- c()
for (i in seq_along(exptsss)) {
    tmp <- substr(e.coli_U00096.2, start = (exptsss[i]-1000), stop = (exptsss[i]+1000))
    if (expstrands[i] == 'forward'){
      expseqs <- c(expseqs, tmp)
    }
    else if (expstrands[i] == 'reverse'){
      expseqs <- c(expseqs, strReverse(tmp))
    }
}

notseqs <- c()
for (i in seq_along(nottsss)) {
    tmp <- substr(e.coli_U00096.2, start = (nottsss[i]-1000), stop = (nottsss[i]+1000))
    if (notstrands[i] == 'forward'){
      notseqs <- c(notseqs, tmp)
    }
    else if (notstrands[i] == 'reverse'){
      notseqs <- c(notseqs, strReverse(tmp))
    }
}

genseqs <- c()
for (i in seq_along(gentsss)) {
    tmp <- substr(e.coli_U00096.2, start = (gentsss[i]-1000), stop = (gentsss[i]+1000))
    if (genstrands[i] == 'forward'){
      genseqs <- c(genseqs, tmp)
    }
    else if (genstrands[i] == 'reverse'){
      genseqs <- c(genseqs, strReverse(tmp))
    }
}


islseqs <- c()
for (i in seq_along(isltsss)) {
    tmp <- substr(e.coli_U00096.2, start = (isltsss[i]-1000), stop = (isltsss[i]+1000))
    if (islstrands[i] == 'forward'){
      islseqs <- c(islseqs, tmp)
    }
    else if (islstrands[i] == 'reverse'){
      islseqs <- c(islseqs, strReverse(tmp))
    }
}



lowscoreseqs <- c()
for (i in seq_along(lowscoretsss)[1:2000]) {
    tmp <- substr(e.coli_U00096.2, start = (lowscoretsss[i]-1000), stop = (lowscoretsss[i]+1000))
    if (lowscorestrands[i] == 'forward'){
      lowscoreseqs <- c(lowscoreseqs, tmp)
    }
    else if (lowscorestrands[i] == 'reverse'){
      lowscoreseqs <- c(lowscoreseqs, strReverse(tmp))
    }
}

#only 2000 lowscore sequences are considered

set.seed(40)
lowscoretsss_short <- sample(lowscoretsss, 2000)
lowscoreseqs_short <- sample(lowscoreseqs, 2000)
```


```{r Sequences statistics}

list_seqs <- grep(pattern = '*seqs', x = ls(), value = T)


list_gcs <- sapply(list_seqs, function(x) {GC(unlist(strsplit(paste0(get(x), collapse = ''), split = '')))})

barplot(list_gcs)

list_gc1 <- sapply(list_seqs, function(x) {GC1(unlist(strsplit(paste0(get(x), collapse = ''), split = '')))})
list_gc2 <- sapply(list_seqs, function(x) {GC2(unlist(strsplit(paste0(get(x), collapse = ''), split = '')))})
list_gc3 <- sapply(list_seqs, function(x) {GC3(unlist(strsplit(paste0(get(x), collapse = ''), split = '')))})

par(mfrow = c(1,5),
    mar = rep(1,4))
for (i in seq_along(list_gc1)) {
  barplot(c(list_gc1[i], list_gc2[i],list_gc3[i]), ylim = c(0.4,.55), main = names(list_gc1[i]))
}

#aug presence

grepl(pattern = 'AGG', get(list_seqs[i]))

library(stringr)

# Count the number of 'a's in each element of string

list_aug_occurences <- list()
for (i in seq_along(list_seqs)) {
  tmp <- sapply(get(list_seqs[i]), function (x) {str_count(x, pattern = 'TATAAT')})
  names(tmp) <- NULL
  list_aug_occurences[[i]] <- tmp
}

par(mfrow = c(1,5))
sapply(list_aug_occurences, function(x) {barplot(unlist(x), ylim = c(0, 85))})

barplot(list_aug_occurences[[1]])
  number.of.AUG <- lapply(get(list_seqs[1]), function (x) {str_count(x, pattern = 'AUG')})
number.of.AUG <- lapply(get(list_seqs[2]), function (x) {str_count(x, pattern = 'AAA')})



#plot(1:nchar(e.coli_U00096.2))
```


```{r Sequences - oligos occurences }

library(Biostrings)


#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(expseqs, notseqs, genseqs, islseqs, lowscoreseqs_short)){
  print(length(i))
  tmp <- DNAString(paste0(unlist(i), collapse = ''))
  
  #tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}



colnames(all_seqs_oligos_freq) <- c('expseqs', 'notseqs', 'genseqs', 'islseqs', 'lowscoreseqs')
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]


par(mfrow = c(3,2))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
  barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

PCA <- prcomp(all_seqs_oligos_freq, scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')

label.size <- 1.7
#library(ggfortify)
#autoplot(PCA,
    #     shape = FALSE,
   #      loadings = TRUE,
  #       label = TRUE,
 #        loadings.label = TRUE,
#         label.size = label.size)+
 # theme_bw()

library(factoextra)

fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=0.6)
  

```


```{r Extracting SIDD profiles for different sequence type from genome-wide vector}

expsidds <- sapply(exptsss, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})

#removing tsss closer than 7500 nts to flanks
nottsss_short <- nottsss[nottsss > 750]
gentsss_short <- gentsss[gentsss > 750]
isltsss_short <- isltsss[isltsss > 750]
lowscoretsss_short <- lowscoretsss[lowscoretsss > 750]

notsidds <- sapply(nottsss_short, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})
gensidds <- sapply(gentsss_short, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})
islsidds <- sapply(isltsss_short, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})
lowscoresidds <- sapply(lowscoretsss_short, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})
islsidds <- sapply(isltsss_short, function(x) {whole_e.coli_sidd_5000_1000[(x-750):(x+750)]})

```

```{r SIDDs set statistics and plots}
L <- grep('*sidds', ls(), value = T)
#par(mfrow = c (2,2)); sapply(L, function(x) {boxplot(get(x))})

#sapply(L, function(x) {apply, 1, sum})

```


```{r circles}
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_2500_500.rda')
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_5000_1000.rda')

maxifind <- function(vec, max){
  v <- vec
  v[which(v<max)] <- 0
  v[which(v>=max)] <- 1
  return(v)
}

maxifind_whole_e.coli_sidd_2500_500 <- maxifind(whole_e.coli_sidd_2500_500, 0.95)
maxifind_whole_e.coli_sidd_5000_1000 <- maxifind(whole_e.coli_sidd_5000_1000, 0.95)

table(maxifind_whole_e.coli_sidd_5000_1000)
table(maxifind_whole_e.coli_sidd_2500_500)


plot(y = rep(0.9, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))), x = which(maxifind_whole_e.coli_sidd_2500_500 ==1),  pch = '|', col = 'red', ylim = c(0,1))
points(y = rep(0.7, length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1))), x = which(maxifind_whole_e.coli_sidd_5000_1000 ==1),  pch = '|', col = 'red')
points(y = rep(0.5, length(exptsss)), x = exptsss,  pch = '|', col = 'darkblue')


#Create data
data = data.frame(
    factor = c(rep('a', length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
              # rep(2, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
               rep('c', length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1)))),
    x = c(which(maxifind_whole_e.coli_sidd_2500_500 ==1),
        #  which(maxifind_whole_e.coli_sidd_2500_500 ==1),
          which(maxifind_whole_e.coli_sidd_5000_1000 ==1)),
    y = c(rep(0.9, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
              # rep(2, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
               rep(0.7, length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1))))
    )
# custom general parameters:
par(mar = c(1, 1, 1, 1), bg=rgb(0.4,0.1,0.7,0.05) ) 
circos.par("track.height" = 0.6)
 
#Now we can initialize the chart. Note that sector.width can be used to custom the size of each sector.

# Step1: Initialize
circos.initialize(factors = data$factor, x = data$x)
circos.trackPlotRegion(y = NULL)
circos.trackPlotRegion(factors = data$factor, y = data$y, panel.fun = function(x, y) {
    circos.axis(
        h="top",                   # x axis on the inner or outer part of the track?
        labels=TRUE,               # show the labels of the axis?
        major.tick=TRUE,           # show ticks?
        labels.cex=0.5,            # labels size (higher=bigger)
        labels.font=1,             # labels font (1, 2, 3 , 4)
        direction="outside",       # ticks point to the outside or inside of the circle ?
        minor.ticks=4,             # Number of minor (=small) ticks
        major.tick.percentage=0.1, # The size of the ticks in percentage of the track height
        lwd=2                      # thickness of ticks and x axis.
        )
    })
 

#The last step fills the region with the chart. You can use all the tradiationnal parameters such as pch, cex, col…

data1 <- subset(data, factor == 'a')
data2 <- subset(data, factor == 'c')


# Step1: Initialise the chart giving factor and x-axis.
circos.initialize( factors=data$factor, x=data$x )
 
# Step 2: Build the regions.
circos.trackPlotRegion(factors = data$factor, y = data$y, panel.fun = function(x, y) {
    circos.axis()
    })
 
# Step 3: Add points
circos.trackPoints(data$factor, data$x, data$y, col = "blue", pch = 1, cex = 0.5)
```



```{r Translate}
library(Biostrings)
library(Peptides)
library(fastcluster)
library(dendextend)

DNA2hydrophob <- function(seq) {
  tmp <- hydrophobicity(as.character(seqinr::translate(strsplit(seq, split = '')[[1]])))
}


hydrophob_expseqs <- lapply(expseqs, DNA2hydrophob)
hydrophob_notseqs <- lapply(notseqs, DNA2hydrophob)


par(mfrow = c(1,2))
boxplot(unlist(hydrophob_expseqs), ylim = c(-5,5))
boxplot(unlist(hydrophob_notseqs), ylim = c(-5,5))

tog <- c(hydrophob_expseqs, hydrophob_notseqs)
table(sapply(tog, length))

tog[which(sapply(tog, function(x) {length(x)<667}))] <- NULL

df <- data.frame(matrix(unlist(tog), nrow=667, byrow=T))

#ward_exp_not_hydrophob <- hclust.vector(X = t(df), method = 'ward')



plot(#ward_exp_not_hydrophob, labels = F)

colvec_exp_not_hydrophob  <- c(rep('darkred', length(expseqs)), rep('darkblue', length(notseqs)))

#ward_exp_not_hydrophob %>% as.dendrogram %>% color_branches(k = length(#ward_exp_not_hydrophob$order), col = colvec_exp_not_hydrophob[order.dendrogram(#ward_exp_not_hydrophob %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(#ward_exp_not_hydrophob$height)*0.2)-> to_plot_#ward_exp_not_hydrophob
```

```{r Translate and Zscale}
library(seqinr)
library(Peptides)
library(fastcluster)
library(dendextend)


zScales_expseqs <- lapply(expseqs, function(x) {sapply(zScales(seqinr::translate(strsplit(x, split = '')[[1]])), FUN = function(x) {x[[1]][1]})})
zScales_notseqs <- lapply(notseqs, function(x) {sapply(zScales(seqinr::translate(strsplit(x, split = '')[[1]])), FUN = function(x) {x[[1]][1]})})

par(mfrow = c(1,2))
boxplot(unlist(zScales_expseqs), ylim = c(-5,5))
boxplot(unlist(zScales_notseqs), ylim = c(-5,5))

tog <- c(zScales_expseqs, zScales_notseqs)
table(sapply(tog, length))

tog[which(sapply(tog, function(x) {length(x)<667}))] <- NULL

df_zsc <- data.frame(matrix(unlist(tog), nrow=667, byrow=T))

#ward_exp_not_zscales<- hclust.vector(X = t(df_zsc), method = 'ward')



plot(#ward_exp_not_zscales, labels = F)

colvec_exp_not_zscales <- c(rep('darkred', length(expseqs)), rep('darkblue', length(notseqs)))

#ward_exp_not_zscales %>% as.dendrogram %>% color_branches(k = length(#ward_exp_not_zscales$order), col = colvec_exp_not_zscales[order.dendrogram(#ward_exp_not_zscales %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(#ward_exp_not_zscales$height)*0.2)-> to_plot_#ward_exp_not_zscales
```

```{r Hydrophobicity of genes with frame shift}

library(seqinr)
library(Peptides)
library(fastcluster)
library(dendextend)

DNA2hydrophob <- function(seq, frame) {
  hydrophobicity(as.character(seqinr::translate(strsplit(seq, split = '')[[1]]), frame = frame))
}


hydrophob_genseqs_frame0 <- lapply(genseqs, function(x) DNA2hydrophob(x, frame = 0))
hydrophob_genseqs_frame1 <- lapply(genseqs, function(x) DNA2hydrophob(x, frame = 1))
hydrophob_genseqs_frame2 <- lapply(genseqs, function(x) DNA2hydrophob(x, frame = 2))


tog <- c(hydrophob_genseqs_frame0, hydrophob_genseqs_frame1, hydrophob_genseqs_frame2)
table(sapply(tog, length))

tog[which(sapply(tog, function(x) {length(x)<667}))] <- NULL

df_frames <- data.frame(matrix(unlist(tog), nrow=667, byrow=T))

#ward_frames_hydrophob <- hclust.vector(X = t(df_frames), method = 'ward')



plot(#ward_frames_hydrophob, labels = F)

colvec_frames_hydrophob  <- c(rep('darkred', length(genseqs)), rep('darkblue', length(genseqs)), rep('darkgreen', length(genseqs)))

#ward_frames_hydrophob %>% as.dendrogram %>% color_branches(k = length(#ward_frames_hydrophob$order), col = colvec_frames_hydrophob[order.dendrogram(#ward_frames_hydrophob %>% as.dendrogram)]) %>% set('labels_cex', 1E-5) %>% raise.dendrogram(max(#ward_frames_hydrophob$height)*0.2)-> to_plot_#ward_frames_hydrophob
```

```
```{r Overall SIDD maxima statistics}

load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_2500_500.rda')
load('/home/jane/Документы/Misha/mol_a_2018/whole_e.coli_sidd_5000_1000.rda')


maxifind <- function(vec, max){
  v <- vec
  v[which(v<max)] <- 0
  v[which(v>=max)] <- 1
  return(v)
}

#maxifind_whole_e.coli_sidd_2500_500 <- maxifind(whole_e.coli_sidd_2500_500, 0.95)
maxifind_whole_e.coli_sidd_5000_1000 <- maxifind(whole_e.coli_sidd_5000_1000, 0.95)

table(maxifind_whole_e.coli_sidd_5000_1000)
#table(maxifind_whole_e.coli_sidd_2500_500)

left_edges <- which(diff(maxifind_whole_e.coli_sidd_5000_1000)==1)
right_edges <- which(diff(maxifind_whole_e.coli_sidd_5000_1000)==-1)

maxima_095_width <- right_edges - left_edges
boxplot(maxima_095_width)


```


```{r Overall maxima width for sequence types and genome}

maxima_position_width <- function(vec, max_cutoff) {
  maxifind <- function(vec, max){
    v <- vec
    v[which(v<max)] <- 0
    v[which(v>=max)] <- 1
    return(v)
  }
  maxifind_tmp <- maxifind(vec, max_cutoff)
  lefts_tmp <- which(diff(maxifind_tmp)==1)
  rights_tmp <- which(diff(maxifind_tmp)==-1)
return(list(lefts_tmp, rights_tmp))
}




```{r Shift in vectors??}
diff(c(length(whole_e.coli_sidd_2500_500), length(whole_e.coli_sidd_5000_1000), nchar(e.coli_U00096.2)))

int <-1: 100000 #maxima are matched
plot(whole_e.coli_sidd_5000_1000[int], type = 'l', ylim = c(-1,1))
lines(-whole_e.coli_sidd_2500_500[int], col = 'red')


int <-100000: 200000 #maxima are matched
plot(whole_e.coli_sidd_5000_1000[int], type = 'l', ylim = c(-1,1))
lines(-whole_e.coli_sidd_2500_500[int], col = 'red')


int <-850000: 900000 #maxima are almost matched
plot(whole_e.coli_sidd_5000_1000[int], type = 'l', ylim = c(-1,1))
lines(-whole_e.coli_sidd_2500_500[int], col = 'red')


#to compare left flank SIDD obtained using inine tool

ecoli_9900 <- substr(e.coli_U00096.2, 1, 9900)
ecoli_9900_sidd_online <- read.table('/home/jane/Документы/Misha/mol_a_2018/9900_left_ecoli_sidd_online.txt', skip = 17, sep = '\t', col.names= c('position', 'P(x)', 'G(x)'))

par(mfrow = c (3,1))
int <-1: 9900 #maxima are matched
plot(whole_e.coli_sidd_5000_1000[int], type = 'l', ylim = c(-1,1))
plot(whole_e.coli_sidd_2500_500[int], type = 'l', ylim = c(-1,1))
plot(ecoli_9900_sidd_online[int,2], type = 'l', ylim = c(-1,1))


lines(-whole_e.coli_sidd_2500_500[int], col = 'red')


```


```{r Sequence analysis for highest maxima of dynamic properties}
load('/home/jane/Документы/Misha/mol_a_2018/df_ecoli_complete_dynamic_chars.rda')
library(Biostrings)
source('/home/jane/Документы/Misha/my_functions/dynchars.R')

whole_e.coli_E01 <- dynchars(strsplit(e.coli_U00096.2, split = '')[[1]], interval_size = 200)$E01
length(whole_e.coli_E01) == nchar(e.coli_U00096.2) #lenghts are checked 

max <- summary(whole_e.coli_E01)['Max.']

maxima_position_width_whole_E01 <- maxima_position_width(whole_e.coli_E01, max_cutoff = max*0.975)

maxima_widths_E01 <- maxima_position_width_whole_E01[[2]] - maxima_position_width_whole_E01[[1]]
which.max(maxima_widths_E01)

midpoints_maxima_E01 <- c()
for (i in seq_along(maxima_position_width_whole_E01[[1]])){
   shift <- round(maxima_widths_E01[i]/2)
   midpoints_maxima_E01 <- c(midpoints_maxima_E01, (maxima_position_width_whole_E01[[2]][i] + shift))
}


#maxima genomic positions
plot(seq_along(whole_e.coli_E01), type = 'n')
abline(v = midpoints_maxima_E01, col = 'red')

maximorum_E01 <- midpoints_maxima_E01[which.max(maxima_widths_E01)]

#half_int <- 500
#plot((maximorum_E01-half_int):(maximorum_E01+half_int), whole_e.coli_E01[(maximorum_E01-half_int):(maximorum_E01+half_int)])

region <- maxima_position_width_whole_E01[[1]][which.max(maxima_widths_E01)]:maxima_position_width_whole_E01[[2]][which.max(maxima_widths_E01)]
plot(region, whole_e.coli_E01[region], type = 'l') #maximorum position

#corresponding sequence

region_sequence <- substr(e.coli_U00096.2, start = min(region), stop = max(region))

region_DNASS <- DNAString(region_sequence)
genome_DNASS <- DNAString(e.coli_U00096.2)


svg('/home/jane/Документы/Misha/mol_a_2018/e.coli_highest_E01_vs_whole_genome_oligomers.svg', width = 10, height = 10)
par(mfrow = c(6,2),
  mar = rep(3.5,4),
  oma = rep(3,4))
for (width in 1:6){
   barplot(sort(oligonucleotideFrequency(region_DNASS, width = width, step=1, as.prob=T), decreasing = T)[1:10], las = 2, main = paste0('Highest E01 region ', width, '-mers'))
   barplot(sort(oligonucleotideFrequency(genome_DNASS, width = width, step=1, as.prob=T), decreasing = T)[1:10], las = 2, main = paste0('Whole genome E01 ', width, '-mers'))
}
 dev.off()

c('A','G')[1*(2*rbinom(1000,1,0.5)/2)]


artificial_seq <- rep(c(rep('A', 13), rep('G', 11), rep('A', 22), rep('G', 12), rep('A', 11), rep('G', 22)), 5)


plot(dynchars(artificial_seq, 200)$E01, type = 'l', ylim  = c(200, 300) )
lines(whole_e.coli_E01[region], col = 'red') #maximorum position

```


```{r Highest E01 points and sequnce}

library()
broadest_maxima <- which(maxima_widths_E01>15) 

broadest_maxima_positions <- c()
broadest_maxima_seqs<- c()
broadest_maxima_E01<- c() #as a sinlge concatenated vector!

for (i in broadest_maxima ){
 broadest_maxima_positions <- cbind(broadest_maxima_positions, c(maxima_position_width_whole_E01[[1]][i], maxima_position_width_whole_E01[[2]][i]))
  broadest_maxima_seqs <- c(broadest_maxima_seqs, substr(e.coli_U00096.2, start = maxima_position_width_whole_E01[[1]][i], stop = maxima_position_width_whole_E01[[2]][i]))
  broadest_maxima_E01<- c(broadest_maxima_E01, whole_e.coli_E01[(maxima_position_width_whole_E01[[1]][i]):(maxima_position_width_whole_E01[[2]][i])])
}

par(mfrow = c (1,2)); boxplot(broadest_maxima_E01, ylim = c(200, 260)); boxplot(whole_e.coli_E01, ylim = c(200, 260))


library(Biostrings)


#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(paste0(broadest_maxima_seqs, collapse = ''), e.coli_U00096.2)){
  print(length(i))
  tmp <- DNAString(paste0(unlist(i), collapse = ''))
  
  #tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}



colnames(all_seqs_oligos_freq) <- c(paste('Highest E01 sequence', nchar(paste0(broadest_maxima_seqs, collapse = '')),'nts'), 'Whole genome')
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]


par(mfrow = c(3,2))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
  barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

PCA <- prcomp(all_seqs_oligos_freq, scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')


library(factoextra)
label.size <- 1.7

fviz_pca_biplot(PCA, 
             axes = c (1,2),
             #labelsize = label.size,
             #alpha.ind="contrib",
             col.ind = 'contrib'
             ) + scale_color_gradient2(low="white", 
                                       mid="blue",
                                       high="red", 
                                       midpoint=1)
  

```


```{r SIDD new fixed whole-genome vectors}

load('/home/jane/Документы/Misha/mol_a_2018/sidd_5000_1000_list_whole_e.coli.rda')


sidd_5000_1000_df <- c()
for (i in sidd_5000_1000_list) {
  #if (nrow(i)<5000) {sidd_5000_1000_df <- rbind(sidd_5000_1000_df, i)
  #} else if (nrow(i)==5000) {
    sidd_5000_1000_df <- rbind(sidd_5000_1000_df, i[-c(4001:5000), ])
  }

nrow(sidd_5000_1000_df) == nchar(e.coli_U00096.2)

over95_positions <- which(sidd_5000_1000_df$P.x.>0.95)
which(diff(which(sidd_5000_1000_df$P.x.>0.95)) !=1)
extended_ecoli_seq <- paste0(tail(sidd_5000_1000_df$Nucleotide, n = 1000),
sidd_5000_1000_df$Nucleotide,
head(sidd_5000_1000_df$Nucleotide, n = 1000))

over95_1500nts_sequences <- sapply(over95_positions, function(x) {extended_ecoli_seq[(x-100+1000):(x+100+1000)]}) #200 nts context over every point with p over 0.95. not optimal

#oligos occurences

library(Biostrings)

#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(paste0(over95_1500nts_sequences, collapse = '')[[1]], e.coli_U00096.2)){
print(length(i))
tmp <- DNAString(paste0(unlist(i), collapse = ''))

#tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}

colnames(all_seqs_oligos_freq) <- c('Max SIDD','genome')
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]

par(mfrow = c(3,2))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

PCA <- prcomp(all_seqs_oligos_freq, scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')

label.size <- 1.7
#library(ggfortify)
#autoplot(PCA,
# shape = FALSE,
# loadings = TRUE,
# label = TRUE,
# loadings.label = TRUE,
# label.size = label.size)+
# theme_bw()

library(factoextra)

fviz_pca_biplot(PCA,
axes = c (1,2),
#labelsize = label.size,
#alpha.ind="contrib",
col.ind = 'contrib'
) + scale_color_gradient2(low="white",
mid="blue",
high="red",
midpoint=0.4)
```


```{r SIDD fro different sequence types}
#extracting one large vector
whole_e.coli_sidd_5000_1000 <- sidd_5000_1000_df$P.x.


#how big are the sets in nts?

lowscoresidds_short <- lowscoresidds[,1:2000]
seqtypes_sidds <- list(whole_e.coli_sidd_5000_1000, expsidds, notsidds, gensidds, islsidds, lowscoresidds_short)
seqtypes_sidds_lengths <- sapply(seqtypes_sidds, function(x) {length(as.numeric(x))})
par(mar = rep(2,4)); barplot(seqtypes_sidds_lengths, names.arg = c('Whole','P', 'N', 'G', 'I', 'L'))

whole_sidds_maxima <- maxima_position_width(whole_e.coli_sidd_5000_1000, 0.95)
whole_sidds_maxima_widths <- whole_sidds_maxima[[2]]-whole_sidds_maxima[[1]]
max(whole_sidds_maxima_widths)
boxplot(whole_sidds_maxima_widths, main = 'Whole genome: \nover 0.95 SIDD regions length')
#for sequence types sets positions are irreleveant
#in order to remove overlapping parts IRanges library is used


library(IRanges)


expIR <- reduce(IRanges(start = exptsss-750, end = exptsss+750))
#expsidds_maxima_widths <- width(expIR)
exp_siddsIR <- list()
for (i in seq_along(start(expIR))){
  exp_siddsIR[[i]] <- whole_e.coli_sidd_5000_1000[start(expIR)[i]:end(expIR)[i]]
}
max(unlist(exp_siddsIR))
expsidds_maxima <- maxima_position_width(unlist(exp_siddsIR), 0.95)
expsidds_maximaIR <- IRanges(start = expsidds_maxima[[1]], end = expsidds_maxima[[2]])


nottsss_trim <- nottsss
nottsss_trim[which(nottsss<750)] <- 750 #flanking sequences are trimed
notIR <- reduce(IRanges(start = nottsss_trim-750, end = nottsss_trim+750))
#notsidds_maxima_widths <- width(notIR)
not_siddsIR <- list()
for (i in seq_along(start(notIR))){
  not_siddsIR[[i]] <- whole_e.coli_sidd_5000_1000[start(notIR)[i]:end(notIR)[i]]
}
max(unlist(not_siddsIR))
notsidds_maxima <- maxima_position_width(unlist(not_siddsIR), 0.95)
notsidds_maximaIR <- IRanges(start = notsidds_maxima[[1]], end = notsidds_maxima[[2]])




gentsss_trim <- gentsss
gentsss_trim[which(gentsss<750)] <- 750 #flanking sequences are trimed
genIR <- reduce(IRanges(start = gentsss_trim-750, end = gentsss_trim+750))
#gensidds_maxima_widths <- width(genIR)
gen_siddsIR <- list()
for (i in seq_along(start(genIR))){
  gen_siddsIR[[i]] <- whole_e.coli_sidd_5000_1000[start(genIR)[i]:end(genIR)[i]]
}
max(unlist(gen_siddsIR))
gensidds_maxima <- maxima_position_width(unlist(gen_siddsIR), 0.95)
gensidds_maximaIR <- IRanges(start = gensidds_maxima[[1]], end = gensidds_maxima[[2]])


isltsss_trim <- isltsss
isltsss_trim[which(isltsss<750)] <- 750 #flanking sequences are trimed
islIR <- reduce(IRanges(start = isltsss_trim-750, end = isltsss_trim+750))
#islsidds_maxima_widths <- width(islIR)
isl_siddsIR <- list()
for (i in seq_along(start(islIR))){
  isl_siddsIR[[i]] <- whole_e.coli_sidd_5000_1000[start(islIR)[i]:end(islIR)[i]]
}
max(unlist(isl_siddsIR))
islsidds_maxima <- maxima_position_width(unlist(isl_siddsIR), 0.95)
islsidds_maximaIR <- IRanges(start = islsidds_maxima[[1]], end = islsidds_maxima[[2]])


lowscoretsss_trim <- lowscoretsss
lowscoretsss_trim[which(lowscoretsss<750)] <- 750 #flanking sequences are trimed
lowscoreIR <- reduce(IRanges(start = lowscoretsss_trim-750, end = lowscoretsss_trim+750))
#lowscoresidds_maxima_widths <- width(lowscoreIR)
lowscore_siddsIR <- list()
for (i in seq_along(start(lowscoreIR))){
  lowscore_siddsIR[[i]] <- whole_e.coli_sidd_5000_1000[start(lowscoreIR)[i]:end(lowscoreIR)[i]]
}
max(unlist(lowscore_siddsIR))
lowscoresidds_maxima <- maxima_position_width(unlist(lowscore_siddsIR), 0.95)
lowscoresidds_maximaIR <- IRanges(start = lowscoresidds_maxima[[1]], end = lowscoresidds_maxima[[2]])


#unify

whole_sidds_maximaIR <- IRanges(start = whole_sidds_maxima[[1]], end = whole_sidds_maxima[[2]])

par(mfrow = c(1,6),
mar= rep(2,4))
lapply(grep('*sidds_maximaIR', ls(), value=T), function(x) {boxplot(get(x)@width, ylim = c(0,100), main = paste0(toupper(substr(x, 0,3)), ',\nover 0.95 width'))})

par(mfrow = c(1,6),
mar= rep(2,4))
lapply(grep('*sidds_maximaIR', ls(), value=T), function(x) {barplot(length(get(x)@width)/sum(get(x)@width), ylim = c(0,0.03), main = paste0(toupper(substr(x, 0,3)), ',\nover 0.95 numbers/length'))})
```

```{r Circos - genomic positions of SIDD maxima and experimental promoters}

maxifind <- function(vec, max){
  v <- vec
  v[which(v<max)] <- 0
  v[which(v>=max)] <- 1
  return(v)
}

maxifind_whole_e.coli_sidd_2500_500 <- maxifind(whole_e.coli_sidd_2500_500, 0.95)
maxifind_whole_e.coli_sidd_5000_1000 <- maxifind(whole_e.coli_sidd_5000_1000, 0.95)

table(maxifind_whole_e.coli_sidd_5000_1000)
table(maxifind_whole_e.coli_sidd_2500_500)

library(circlize)

#SIDD maxima and experimental TSSs do not match
region <- 1:5000
#plot(y = rep(0.9, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))), x = which(maxifind_whole_e.coli_sidd_2500_500 ==1),  pch = '|', col = 'red', ylim = c(0,1))
##plot(region, type = 'n',ylim = c(0,1))
plot(y = rep(0.7, length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1)))[region], x = which(maxifind_whole_e.coli_sidd_5000_1000 ==1)[region],  pch = '|', col = 'red')
points(y = rep(0.65, length(exptsss))[region], x = exptsss[region],  pch = '|', col = 'darkblue')


#Create data
data = data.frame(
    factor = c(rep('a', length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
              # rep(2, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
               rep('c', length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1)))),
    x = c(which(maxifind_whole_e.coli_sidd_2500_500 ==1),
        #  which(maxifind_whole_e.coli_sidd_2500_500 ==1),
          which(maxifind_whole_e.coli_sidd_5000_1000 ==1)),
    y = c(rep(0.9, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
              # rep(2, length(which(maxifind_whole_e.coli_sidd_2500_500 ==1))),
               rep(0.7, length(which(maxifind_whole_e.coli_sidd_5000_1000 ==1))))
    )
# custom general parameters:
par(mar = c(1, 1, 1, 1), bg=rgb(0.4,0.1,0.7,0.05) ) 
circos.par("track.height" = 0.6)
 
#Now we can initialize the chart. Note that sector.width can be used to custom the size of each sector.

# Step1: Initialize
circos.initialize(factors = data$factor, x = data$x)
circos.trackPlotRegion(y = NULL)
circos.trackPlotRegion(factors = data$factor, y = data$y, panel.fun = function(x, y) {
    circos.axis(
        h="top",                   # x axis on the inner or outer part of the track?
        labels=TRUE,               # show the labels of the axis?
        major.tick=TRUE,           # show ticks?
        labels.cex=0.5,            # labels size (higher=bigger)
        labels.font=1,             # labels font (1, 2, 3 , 4)
        direction="outside",       # ticks point to the outside or inside of the circle ?
        minor.ticks=4,             # Number of minor (=small) ticks
        major.tick.percentage=0.1, # The size of the ticks in percentage of the track height
        lwd=2                      # thickness of ticks and x axis.
        )
    })
 

#The last step fills the region with the chart. You can use all the tradiationnal parameters such as pch, cex, col…

# Step 3: Add points
circos.trackPoints(data$factor, data$x, data$y, col = "blue", pch = 1, cex = 0.5)
```

```{r Sequence of precise region of SIDD over 0.95}

over95_positions <- which(sidd_5000_1000_df$P.x.>0.95)
over95_precise_sequences <- as.character(sidd_5000_1000_df$Nucleotide[which(sidd_5000_1000_df$P.x.>0.95)])


#oligos occurences

library(Biostrings)

#better way to find oligos frequnces with biostrigs
all_seqs_oligos_freq <- c()
for (i in list(paste0(over95_precise_sequences, collapse = '')[[1]], e.coli_U00096.2)){
print(length(i))
tmp <- DNAString(paste0(unlist(i), collapse = ''))

#tmp_StringSet <- DNAStringSet(sapply(t(tmp), function(x) {paste0(x, collapse = '')}))

all_seqs_oligos_freq <- cbind(all_seqs_oligos_freq, oligonucleotideFrequency(tmp, 6, as.prob = T))
}

colnames(all_seqs_oligos_freq) <- c('Precise max SIDD','genome')
View(all_seqs_oligos_freq)
#phrasetables_6 <- lapply(grep('*_ng6', x = ls(), value = T), FUN = function(x) {get.phrasetable(get(x))}) #no [1:20,]

par(mfrow = c(2,1))
ylim = range(all_seqs_oligos_freq)

for (i in 1:ncol(all_seqs_oligos_freq)){
barplot(sort(all_seqs_oligos_freq[1:20,i], decreasing = T), las = 2, ylim = ylim, main = colnames(all_seqs_oligos_freq) [i])
}

PCA <- prcomp(all_seqs_oligos_freq, scale = T)
#library(factoextra)
#fviz_pca_biplot(PCA12, col.var = 2, labelsize = 0.8)+ggtitle('PCA result on hexamers occurences')



library(factoextra)
label.size <- 1.7

fviz_pca_biplot(PCA,
axes = c (1,2),
#labelsize = label.size,
#alpha.ind="contrib",
col.ind = 'contrib'
) + scale_color_gradient2(low="white",
mid="blue",
high="red",
midpoint=0.4)

library(ape)
GC.content(as.DNAbin(over95_precise_sequences))

```


